---
title: "Phylogenetic Diversity - Communities"
author: "Z620: Quantitative Biodiversity, Indiana University"
date: "February 26, 2015"
header-includes:
   - \usepackage{array}
output: pdf_document
geometry: margin=2.54cm
---

## OVERVIEW
Taxonomic measures of $\alpha$- and $\beta$-diversity do not integrate evolutionary information. 
However, evolutionary information is imortant for broad range of biodiversity issues including conservation biology, biogeography, and community assembly. 
In this handout, we will introduce some commonly used methods in phylogenetic community ecology.
These methods will allow us to account for phylogenetic diversity and provide insight into the mechansisms that give rise to over- and under-dispersion (i.e., clustering) of biological communities. 

After completing this exercise you will know how to:

1.  measure phylogenetic $\alpha$ diversity
2.  measure phylogenetic $\beta$ diversity
3.  evaluate the contribution of phylogenetic information to geographical patterns of biodiversity

## 1) SETUP
### A. Retrieve and Set Your Working Directory

```{r, results = 'hide'}
rm(list = ls())
getwd()
setwd("~/GitHub/QuantitativeBiodiversity/Assignments/PhyloCom")
```

### B. Load Packages 
We will be relying heavily on the R package `picante`. 
This package has many of the functions that are contained in the software Phylocom, which is used for the analysis of phylogenetic community structure and character evolution (http://phylodiversity.net/phylocom/).
We will also use a couple of packages that were introduced in the Phylogenetic Traits module. 

After the initial installation of these packages using the `install.packages()` fucntion, let's load the packages and their dependencies with the `require()` function:

```{r, results = 'hide', message = FALSE, warning = FALSE} 
require("picante")
require("ape")
require("seqinr")
```

### C. Load Source Code
In addition to relying on contributed packages, this week we will also be using a source code file.
A source code file has user-defined functions that are required for certain analyses.
The benefit of source files is that they contain "vetted" code that can be used across multiple projects. 
Here, we will be using a source code file that includes a function for reading in the output files from the popular community sequencing software `mothur` (http://www.mothur.org/).

```{r}
source("./bin/MothurTools.R")
```

## 2) DESCRIPTION OF DATA
We will revisit the data that was used in the Geographical Ecology module. 
As a reminder, in 2013 we sampled ~ 50 forested ponds located in Brown County State Park, Yellowood State Park, and Hoosier National Forest. 
In addition to measuring a suite of geographic and environmental variables, we characterized the diversity of bacteria in the ponds using molecular-based approaches.
Specifically, we amplified the 16S rRNA gene (i.e., "DNA") and 16S rRNA transcripts (i.e., "RNA") of bacteria using barcoded primers on the Illumina MiSeq platform.
We then used a `mothur` pipeline to quality-trim our data set and assign sequences to operational taxonomic units (OTU).
In this exercise, we will use the DNA sequence data for making community phylogenetic inferece.  

## 3) LOADING OF DATA
First, let us load the environmental data:

```{r}
# Load Environmental Data
env <- read.table("data/20130801_PondDataMod.csv", sep = ",", header = TRUE)
```

Next, let us load the bacterial OTU data (i.e., site-by-species matrix):

```{r}
# Load Site-by-Species Matrix
comm <- read.otu(shared = "./data/INPonds.final.rdp.shared", cutoff = "1")

# Select DNA Data: Use `grep()` Command and Rename with `gsub()`
comm <- comm[grep("*-DNA", rownames(comm)), ]
rownames(comm) <- gsub("\\-DNA", "", rownames(comm))
rownames(comm) <- gsub("\\_", "", rownames(comm))

# Remove Sites Not in Environmental Data Set
comm <- comm[rownames(comm)  %in% env$Sample_ID, ]

# Remove Zero-Occurrence Taxa 
comm <- comm[ , colSums(comm) > 0]

# Import Taxonomy Data Using `read.tax()` from Source Code
tax <- read.tax(taxonomy = "./data/INPonds.final.rdp.1.cons.taxonomy")
```

Now, let's load and process some of the phylogenetic data:

```{r}
# Import the Alignment File {seqinr}
ponds.cons <- read.alignment(file = "./data/INPonds.final.rdp.1.rep.fasta", format = "fasta")  

# Rename OTUs in FASTA File
ponds.cons$nam <- gsub("\\|.*$", "", gsub("^.*?\t", "", ponds.cons$nam))

# Import Outgroup FASTA File {seqinr}
outgroup <- read.alignment(file = "./data/methanosarcina.fasta", format = "fasta")

# Convert Alignment File to DNAbin Object {ape}
DNAbin <- rbind(as.DNAbin(outgroup), as.DNAbin(ponds.cons))

# Visusalize Sequence Alignment {ape}
image.DNAbin(DNAbin, show.labels=T, cex.lab = 0.05, las=1) 

# Create Distance Matrix with the Jukes Cantor "JC" Model {ape}
seq.dist.jc <- dist.dna(DNAbin, model = "JC", pairwise.deletion = FALSE)

# Use Neighbor Joining Algorithm to Construct a Full Tree (DNA and RNA sequences) {ape}
phy.all <- bionj(seq.dist.jc)

# Drop Tips of Zero-Occurrence OTUs (Removes Taxa Only Found via RNA sequences) {ape}
phy <- drop.tip(phy.all, phy.all$tip.label[!phy.all$tip.label %in% 
                                             c(colnames(comm), "Methanosarcina")])

# Identify Outgroup Sequence
outgroup <- match("Methanosarcina", phy$tip.label)

# Root the Tree {ape}
phy <- root(phy, outgroup, resolve.root = TRUE)

# Plot the Rooted Tree {ape}
par(mar = c(1,1,2,1) + 0.1)
plot.phylo(phy, main = "Neighbor Joining Tree", "phylogram", 
           use.edge.length = FALSE, direction = "right", cex = 0.6, label.offset = 1)
```

## 4) PHYLOGENETIC ALPHA DIVERSITY

### A.  Faith's Phylogenetic Diversity (PD)
In 1992, Daniel Faith developed a diversity metric called Faith's PD (http://goo.gl/wM08Oy).
The metric sums the branch lengths for each species found in a sample from the root to the tip of a the phylogenetic tree.
The value of the metric captures the evolutionary history of an assemblage. 
Higher PD values indicate that an assemblage contains more evoluiotnarily divergent taxa, while lower PD values indicate that an assemblage contains taxa with a more restricted evolutionary history. 

Faith's PD can be implemented in R using the `pd()` function in the `picante` package. 
A phylognetic tree containing the species pool is required. 
In addition to returning Faith's PD, the `pd()` function also returns species richness (SR).
SR is the same as observed richness ($\S_{obs}), which we covered in the $\alpha$ diversity module. 

```{r}
# Calculate PD and SR {picante}
pd <- pd(comm, phy, include.root = FALSE)
```

Let's compare PD estimates with SR of our samples.
We'll use a (natural) log-log transformation, so that the slope of the relationship gives us a power-law exponent, which describes how PD scales with SR.

```{r}
# Biplot of SR and PD
par(mar = c(5, 5, 4, 1) + 0.1)

plot(log(pd$SR), log(pd$PD), 
     pch = 20, col = "red", las = 1,
     xlab = "ln(SR)", ylab = "ln(PD)", cex.main = 1,
     main="Phylodiversity (PD) vs. Taxonomic richness (SR)")

fit <- lm('log(pd$PD) ~ log(pd$SR)')
abline(fit, col = "red", lw = 2)
exponent <- round(coefficients(fit)[2], 2)
legend("topleft", legend=paste("Scaling exponent = ", exponent, sep = ""), 
       bty = "n", lw = 2, col = "red") 
```

***Question 1***: Answer the following questions about the PD-SR pattern. 
a.  Describe the relationship between taxonomic richness and phylodiversity?
b.  Mathematically, why should they be correlated?  
c.  Under what conditions would you expect these two estimates of diversity to deviate from one another?

> ***Answer 1a***:  
> ***Answer 1b***:  
> ***Answer 1c***:  

#### i.  Randomizations and Null Models
Randomizations are a way of resampling data to assess whether or not observed patterns are different from a null expectation. 
A number of the functions in `picante` allow us to specify different null models as an argument. 
These null models can control for features such as species richness, species occurrence frequencey, and the diversity of the regional species pool. 
We will use some of these models for assesing the degree to which phylogenetic measures of $\alpha$ diversity deviate from null expecations. 
The following table describes some of the null models that are available to us when using `picante`:

\begin{center}
\hyphenpenalty 10000
\exhyphenpenalty 10000
\begin{tabular}{ m{5cm}  m{10cm} }
  \textbf{Null Model} & \textbf{Description} \\
  \hline \hline \\ [-1.5ex]
  \textbf{taxa.labels} & 
  Shuffles taxa labels across tips of phylogeny (across all taxa included in phylogeny) \\
  \\ [-1.5ex]
  \textbf{richness} & 
  Randomizes community data matrix abundances within samples (maintains sample species richness) \\
  \\ [-1.5ex]  
  \textbf{frequency} & 
  Randomizes community data matrix abundances within species (maintains species occurence frequency) \\
  \\ [-1.5ex]   
  \textbf{sample.pool} & 
  Randomizes community data matrix by drawing species from pool of species occurring in at least one community (sample pool) with equal probability \\
  \\ [-1.5ex]  
    \textbf{phylogeny.pool} & 
  Randomize community data matrix by drawing species from pool of species occurring in at least one community (sample pool) with equal probability \\
  \\ [-1.5ex]
  \textbf{independentswap} & 
  Randomizes community data matrix with the independent swap algorithm (Gotelli 2000) maintaining species occurrence frequency and sample species richness \\
  \\ [-1.5ex]
  \textbf{trialswap} & 
  Randomizes community data matrix with the trial-swap algorithm (Miklos \& Podani 2004) maintaining species occurrence frequency and sample species richness \\
  \\ [-1.5ex]
    \hline
\end{tabular}
\end{center}

Now, we are going to use the `ses.pd()` function in `picante`.
This function estimates the standardized effect size ("ses") using the following equation: `ses.pd` = (`pd.obs` - `pd.rand.mean`) / `pd.rand.sd`, where `pd.obs` is the observed PD, `pd.rand.mean` is the mean of the PD values generated via randomization under a null model, and `pd.rand.sd` is the standard deviation of the PD values generated via randomization under a null model (see table above). 
Given the size of both our site-by-species matrix and the phylogenetic tree, the randomizaiton process is computationally intensive. 
Therfore, we are only going to run the `ses.pd` function for two ponds with a limited number of randomizations (i.e., "runs" argument).

```{r}
# Estimate Standardized Effect Size of PD via Randomization {picante}
ses.pd <- ses.pd(comm[1:2,], phy, null.model = "richness", runs = 25, include.root = FALSE)
```

***Question 2***: Using `help()` and the table above, identify two null models that can be used with the `ses.pd()` function.
Rerun `ses.pd()` with these null models and answer the following questions:

a.  What hypotheses are being tested with the p-values associated with `ses.pd`?
b.  What features might affect the intepretation of the `ses.pd` output?

> ***Answer 2a***:  
> ***Answer 2b***:  

### B.  Phylogenetic Dispersion Within a Sample
Another way to assess the phylogenetic $/alpha$-diversity is to look at dispersion within a sample. 
In the following section we will introduce to commonly used metrics --- the Net Relatedness Index (NRI) and the Nearest Taxon Index --- to quantify the degree to which closely related taxa co-occur.
We will use randomization procedures to test whether species are phylogenetically clustered or overdispersed. 

#### i. Phylogenetic Resemblance Matrix 
Before estimating dispersion metrics, we need to create a phylogenetic resemblance matrix.
This type of matrix is nearly identical to the resemblance matrix introduced in the $/beta$-diversity module.
The only difference is that the phylogenetic resemblance matrix contains distances between taxa in a tree, whereas the community resemblance matrix contains distances among sites. 
The elements in phylogenetic resemblance matrix are calculated as the pairwise branch-length distances between tips (i.e., taxa) on a phylogenetic tree. 
The phylogenetic resemblance matrix is sometimes referred to as the phylogenetic variance-covariance matrix. 
We will use the `cophenetic.phylo()` function in `picante` to calculate the phylogenetic resemblance matrix.

```{r}
# Create a Phylogenetic Distance Matrix {picante}
phydist <- cophenetic.phylo(phy)
```

#### ii.  Net Relatedness Index (NRI)
One common way to test for phylogenetic clustering and overdispersion is to use the Net Relatedness Index (NRI).
NRI is based on quantifying the mean phylogenetic distance (MPD).
MPD is the mean phylogenetic distance from pairwise branch lengths in a sample.
With this information in hand, NRI is expressed as: - (`mpd.obs` - `mpd.rand.mean`) / `mpd.rand.sd` where `mpd.obs` is the observed MPD, `mpd.rand.mean` is the mean of the MPD values generated via randomization under a null model, and `mpd.rand.sd` is the standard deviation of the MPD values generated via randomization under a null model. 

Negative NRI values indicate that a sample is phylogenetically overdispersed; that is, taxa are less related to one another than expected under the null model.
Positive NRI values indicate that a sample is phylogentically underdispersed, or clustered, such that taxa are more closely related to one another than expected under the null model. 

As with Faith's PD, the radomization procedures are computationally intensive, so we are only going to peform a relatively small number of "runs".

```{r}
# Estimate Standardized Effect Size of NRI via Randomization {picante} 
ses.mpd <- ses.mpd(comm, phydist, null.model = "taxa.labels", abundance.weighted = FALSE, runs = 25) 

# Calculate NRI
NRI <- as.matrix(-1 * ((ses.mpd[,2] - ses.mpd[,3]) / ses.mpd[,4]))
rownames(NRI) <- row.names(ses.mpd)
colnames(NRI) <- "NRI"
```

### iii.  Nearest Taxon Index (NTI)
Another way to test for phylogenetic clustering and oversdispersion in a sample is to use the Nearest Taxon Index (NTI).
This index is mathemically similar to NRI, but uses the mean nearest phylogenetic neighbor distance (MNND) insted of MPD. 
MNND is the mean phylogenetic distance between all taxa in a sample and their phylogenetically closest neighbor.
As a result, NTI tends to emphasize terminal clustering, independent of deep level clustering (Webb et al. 2002; http://goo.gl/WikgWE).
Just like NRI, we peform randomizations and use this information to estimate the standardized effect size. 
Negative NRI values indicate phylogenetic overdispersion and positive NRI values indicate phylogenetic clustering. 

```{r}
# Estimate Standardized Effect Size of NRI via Randomization {picante} 
ses.mntd <- ses.mntd(comm, phydist, null.model = "taxa.labels", abundance.weighted = FALSE, runs = 25)

# Calculate NTI
NTI <- as.matrix(-1 * ((ses.mntd[,2] - ses.mntd[,3]) / ses.mntd[,4]))
rownames(NTI) <- row.names(ses.mntd)
colnames(NTI) <- "NTI"
```

***Question 3***: In the NRI and NTI examples above, the arguments "abundance.weighted = FALSE" means that the indices were calculated using presence-absence data.
Modify and rerun the code so that NRI and NTI are caculated using abundance data. 
How does this affect the interpretation of NRI and NTI?

> ***Answer 3***:  

### C) Mini-Exercise On Phylogenetic Alpha-Diversity

## POTENTIAL EXERCISE: HAVE STUDENTS LOOK FOR CORRELATIONS BETWEEN NRI/NTI AND ENVIRONMENTAL VARIABLES. 
## DOES CLUSTERING/DISPERSION CHANGE ALONG GRADIENTS OR WITH PD? IF SO, WHAT ARE SOME DIFFERENT EXPLANATIONS?

## 5) PHYLOGENETIC BETA DIVERSITY 

### A. A Phylogenetically Based Community Resemblance Matrix
Just as we needed a resemblance matrix in taxonomic $\beta$-diversity, we need a community resemblace matix to quantifiy phylogenetic $\beta$-diversity. 
However, instead of making the resemblance matrix based on incidence or abundance of taxa we are now going to incorporate information about the phylogenetic relationships among taxa. 
Similar to other measures of $\beta$ diversity, there are numerous ways to calculate the phylogenetic distances in the community resemblance matrix.

\begin{center}
\hyphenpenalty 10000
\exhyphenpenalty 10000
\begin{tabular}{ m{5cm}  m{10cm} }
  \textbf{Index} & \textbf{Description} \\
  \hline \hline \\ [-1.5ex]
  \textbf{Mean Pairwise Distance} & 
  The phylogenetic distance between two taxa: the sum of all intervening branch lengths. \\
  \\ [-1.5ex]
  \textbf{UniFrac} & 
  The distance between the two samples calculated as (the sum of "unshared" branch lengths)/(the sum of all tree branch lengths) in a rooted tree. \\
  \\ [-1.5ex]
  \hline
\end{tabular}
\end{center}

```{r}
# Mean Pairwise Distance
dist.mp <- comdist(comm, phydist)

# UniFrac Distanc
dist.uf <- unifrac(comm, phy)
```

Here we are calcuating both as unweighted distances (similar to incidence based measures).
That means that we are not taking into account the abundance of each taxa in each site.
Here are, however, weighted versions of both calculations (similar to abundance based measurse).

### B. Visualizing Phylogenetic Beta Diversity

Now that we have our Phylogenetic Resemblance Matrix, we can visualize $\beta$ diversity using the same techniques we used in the $\beta$-diversity module.
Here we are just going to use ordination, but any of the viualization techniques would work. 
Here, we will use Principal Coordinates Analysis (PCoA), using the UniFrac distance based resemblance matrix.
We will use the `cmdscale()` function to conduct our PCoA analysis.
Additinally, we we calculate our explained variation on each axis.

```{r}
pond.pcoa <- cmdscale(dist.uf, eig = T, k = 3)

explainvar1 <- round(pond.pcoa$eig[1] / sum(pond.pcoa$eig), 3) * 100
explainvar2 <- round(pond.pcoa$eig[2] / sum(pond.pcoa$eig), 3) * 100
explainvar3 <- round(pond.pcoa$eig[3] / sum(pond.pcoa$eig), 3) * 100
sum.eig <- sum(explainvar1, explainvar2, explainvar3)
```

Now that we have calculated our PCoA, we can plot the results.
Remember, you should check the eigenvalues to determine your confidence in the data reduction approach.

```{r pcoalayer1, eval = TRUE, fig.width=6, fig.height=4}
# Define Plot Parameters
par(mar = c(5, 5, 1, 2) + 0.1)

# Initiate Plot
plot(pond.pcoa$points[ ,1], pond.pcoa$points[ ,2],
     xlim = c(-0.2, 0.2), ylim = c(-0.25, 0.15),
     xlab = paste("PCoA 1 (", explainvar1, "%)", sep = ""),
     ylab = paste("PCoA 2 (", explainvar2, "%)", sep = ""),
     pch = 16, cex = 2.0, type = "n", cex.lab = 1.5, cex.axis = 1.2, axes = FALSE)

# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
abline(h = 0, v = 0, lty = 3)
box(lwd = 2)

# Add Points & Labels
points(pond.pcoa$points[ ,1], pond.pcoa$points[ ,2],
       pch = 19, cex = 3, bg = "gray", col = "gray")
text(pond.pcoa$points[ ,1], pond.pcoa$points[ ,2], 
     labels = row.names(pond.pcoa$points))
```

### C. Hypothesis Testing

#### i. Categorical Approach: Water Shet

```{r}
# Define Environmental Catagory
water.shed <- env$Location


# Run PERMANOVA with adonis function
adonis(dist.uf ~ water.shed, permutations = 999)
```

#### ii. Continuous Approach: Environmental Gradients

```{r}
# Define Environmental Gradient
envs <- env[,5:19]



```



 phyloordination ?Phylogenetic beta diversity measures can be used with any method based onvmeasuring among-community distances. For example, they could be used in a cluster analysis or phyloordination to group communities based on their evolutionaryvsimilarity, or they could be compared with spatial or environmental distances separating communities using a Mantel test. The code below calculates MPD between pairs of communities, and uses these phylogenetic distances to cluster communities based on their phylogenetic similarity:


Ordinations: PCoA taxo  vs. UNIFRAC
http://www.inside-r.org/packages/cran/picante/docs/unifrac
http://cran.r-project.org/web/packages/GUniFrac/GUniFrac.pdf
http://joey711.github.io/phyloseq-demo/unifrac.html  <- out of date. not worth it

```{r}

unifrac(comm, phydist)
# GUniFrac(comm, phydist)   # Tree Must be rooted


GUniFrac(otu.tab, tree, alpha = c(0, 0.5, 1))

# Calculate the UniFracs
unifracs <- GUniFrac(otu.tab.rff, throat.tree, alpha=c(0, 0.5, 1))$unifracs

#  Combine unweighted and weighted UniFrac for testing
PermanovaG(unifracs[, , c("d_1", "d_UW")] ~ groups)
# Combine d(0), d(0.5), d(1) for testing
PermanovaG(unifracs[, , c("d_0", "d_0.5", "d_1")] ~ groups)



phylosor.rnd(samp,tree, cstSor=TRUE, null.model=c("taxa.labels",
    "frequency","richness","independentswap","trialswap"),
    runs=999, iterations=1000)

```


## 6) PHYLOGENETIC DIVERSITY-AREA RELATIONSHIP (PDAR)
In this week's reading, you learned how Helmus and Ives (2012) studied how phylogenetic diversity changes with increasing area.
This phylogenetic diversity-area relationship (PDAR) is analogous to the species-area relationship (SAR) that we learned about in Geographical Ecology.
Remember that the SAR is a cumulative relationship (S = cA^z) and so, cannot be negative.
In fact, the shallowest the SAR can get is to have a slope of 0.0, meaning that all species are found in all samples.
In contrast the PDAR can increase or decrease with area, which provoked Helmus and Ives (2012) to develop theoretical expectations for the effects of forces or processes on the shape of the PDAR.

Helmus and Ives use the Phylogenetic species variability (PSV) metric to quantify phylogenetic diversity.
PSV quantifies how phylogenetic relatedness decreases the variance of a hypothetical neutral trait shared by all species in a community. 
Type 'help(psv)' to learn more.

In their study, Helmus and Ives, increased area by aggregating plots at random either with or without respect to whether the plots were adjacent.
Both sampling methods produce roughly the same result.
Here, we will construct PDARs using the random aggregation approach of Helmus and Ives (2012).
**Helmus and Ives (2012) did this in a way that is similar to how we constructed SARs in the Geographical Ecology lesson.**

Let's begin by writing a function to generate the PDAR.

```{r}
PDAR <- function(comm, tree){
  # Use the PSV index of Helmus and Ives (2012) by default
  # Using PD.index, we've left room to call other diversity metrics
  
  areas <- c() # an object to hold areas
  diversity <- c() # an object hold diversity
  
  num.plots <- c(2,4,8,16,32,51) # increase numbers of plots by 2, roughly doubling area
  
  for (i in num.plots){    
    areas.iter <- c() # hold areas from iterations; we'll find the mean from this
    diversity.iter <- c() # hold estimated diversity from iterations; we'll find the mean
    
    for (j in 1:10){ # Iterate 100 times for each sample size
      pond.sample <- sample(51, replace = FALSE, size = i) # sampling w/out replacement
      area <- 0 # a variable to hold accumulating area
      sites <- c() # a vector to hold accumulating taxa
      
      for (k in pond.sample) { # Loop through each randomly drawn pond...  
        area <- area + pond.areas[k] # aggregating area...
        sites <- rbind(sites, comm[k, ]) # and sites
        }
      
      areas.iter <- c(areas.iter, area) # concatenate the area to areas.iter      
      psv.vals <- psv(sites, tree, compute.var=FALSE) # calculate PSV
      psv <- psv.vals$PSVs[1] 
      diversity.iter <- c(diversity.iter, as.numeric(psv))
      }

    diversity <- c(diversity, mean(diversity.iter)) # let diversity be the mean PSV
    areas <- c(areas, mean(areas.iter)) # let area be the average area
    print(c(i, mean(diversity.iter), mean(areas.iter))) # print as we go
    }
  return(cbind(areas, diversity)) # return vectors of areas (x) and diversity (y)
  }
```

Now, let's examine the PDAR for our pond data set.
We will examine the relationship of phylogenetic diversity to area using both Spearman's correlation coefficient (S) and Pearson's correlation coefficient (P).
It is informative to use both because while S is computed on ranks and depicts monotonic relationships (the degree to which the relationship is continually increasing or decreasing), P is computed on the observed values and therefore depicts linear relationships.
For example, if you S > P then the correlation is more monotonic than linear and vice versa.

```{r}
#par(mfrow=c(1, 2))
pdar <- PDAR(comm, phy) # Compute the PDAR
pdar <- as.data.frame(pdar)

Pearson <- cor.test(pdar$areas, pdar$diversity, method="pearson")
P <- round(Pearson$estimate,2)
Pp <- round(Pearson$p.value,3)

Spearman <- cor.test(pdar$areas, pdar$diversity, method="spearman")
S <- round(Spearman$estimate,2)
Sp <- round(Spearman$p.value,3)

plot(pdar, xlab='Area', ylab='PSV', main='Phylogenetic diversity-area relationship', col='red')
legend("topleft", legend=paste("Spearman correlation =", S, "p =", Sp,
                  "\nPearson correlation =", P, "p =", Pp), bty="n", lw=0, col="red") 
```


Now, let's write a function to generate the SAR, similar to how we could have done it a few weeks ago.

```{r}
SAR <- function(OTUs){
  Alist <- c()
  Slist <- c()
  
  num.ponds <- c(2,4,8,16,32,51)
  for (i in num.ponds) {   
    areas <- c() # hold iterated area values 
    Ss <- c() # hold iterated S values
    
    for(j in 1:100){
      pond.sample <- sample(51, replace = FALSE, size = i) 
      area <- 0
      cum.abs <- vector(mode="numeric", length=length(comm))
  
      for (k in pond.sample) { # Loop through each randomly drawn pond
        area <- area + pond.areas[k] # aggregating area, (incrementing)
        cum.abs <- cum.abs + as.numeric(comm[k, ])
    
      } # End random pond samples loop
      S <- length(cum.abs[cum.abs > 0])
      areas <- c(areas, area)
      Ss <- c(Ss, S)
    }
    Alist <- rbind(Alist, mean(areas))
    Slist <- rbind(Slist, mean(Ss))
    print(c(mean(areas), mean(Ss)))
  }
  return(cbind(log10(Alist), log10(Slist)))
}
```

Now, let's generate the SAR for our ponds dataset.

```{r}
pond.areas <- as.vector(pi * (env$Diameter/2)^2) # Find areas of all 51 ponds
sar <- SAR(comm)
sar <- as.data.frame(sar)

plot(sar, xlab="log(Area)", ylab="log(Richness)", 
     main="Species-area relationship", col="red")

OLS <- lm('sar$V2 ~ sar$V1')
abline(OLS, col = "red", lw = 2)

slope <- round(coefficients(OLS)[2], 3)
legend("topleft", legend = paste("slope =", slope), bty = "n", lw = 2, col = "red")  
```


***Question X**: We observed a slope for our SAR that is only slightly lower than would normally be expected, but which is steeper than for many microbial communities. However, what did we observe for the PDAR and how might we explain the differences in the results of these biodiversity patterns, one being taxonomic (SAR) and the other being phylogenetic (PDAR)?

> ***Answer X***:


## 6) PHYLOGENETIC DISTANCE-DECAY RELATIONSHIP (PDD)
Remember from the Geographical Ecology lesson that the Distance Decay relationship (DD) reflects the spatial autocorrelation of community similarity.
That is, geographically near communities should be more similar than geographically distant communities.
Given that similarity in community composition generally decreases with geographic distance, let's apply the concept of spatial autocorrelation to phylogenetic diversity and ask whether similarity in phylogenetic diversity also decreases with distance.
Instead of the PSV metric of Helmus and Ives (2012), we'll use the UNIFRAC distance, which is more analogous to the Bray-Curtis metric used in the DD.

To generate the PDD, we will need to calculate geographic distances based on lat-long coordinates.
Let's use the `fossil` package for this.
```{r}
require("fossil")
require("simba")
```

Next, let's write a function to find the geographic distances and the differences in phylogenetic diversity between all unique pair-wise combinations of ponds.

```{r}
unifrac.dist <- unifrac(comm, phy) # unifrac distance
bray.curtis.dist <- 1 - vegdist(OTUs) # Bray-Curtis similarity between the plots
coord.dist <- dist(as.matrix(env$lat, env$long)) # geographic distance between plots
  
# transform all distance matrices into list format:
unifrac.dist.ls <- liste(unifrac.dist, entry="unifrac")
bray.curtis.dist.ls <- liste(bray.curtis.dist, entry="bray.curtis")
coord.dist.ls <- liste(coord.dist, entry="geo.dist")

# Now, create a data frame containing similarity of the environment and similarity of community.
df <- data.frame(coord.dist.ls, bray.curtis.dist.ls[,3], unifrac.dist.ls[,3])
names(df)[4:5] <- c("bray.curtis", "unifrac")
attach(df) #df <- subset(df, struc != 0)

# Finally, let's plot the Distance-decay relationships, with regression lines in red.
plot(geo.dist, bray.curtis, xlab="Geographic Distance", ylab="Bray-Curtis Similarity", 
   main = "Distance-Decay for Taxa", col='SteelBlue')

OLS <- lm(env ~ dist)
OLS # print regression results to the screen
abline(OLS, col="red4")

par(mfrow=c(1, 1))
plot(geo.dist, unifrac, xlab="Geographic Distance", ylab="Unifrac Distance", 
     main="Distance-Decay for Phylogenetic", col='darkorchid4')

OLS <- lm(struc ~ dist)
OLS # print regression results to the screen
abline(OLS, col="red4")

# Let's, examine the slope of the regression lines, asking whether they are significantly different from one another.

diffslope(geo.dist, unifrac, geo.dist, bray.curtis)
```


## 7) HOMEWORK 

1.  In their study of phylogenetic diversity-area relationships (PDARs), Helmus and Ives, increased area by aggregating plots in two ways.
First, they aggregated plots with respect to whether the plots were adjacent; what they called 'spatial'.
Second, they aggregated plots at random ('non-spatial').
While both spatial and non-spatial sampling methods capture the effect of increasing area, sampling with respect to location (i.e. spatial) also captures the effect of increasing distance or spatial autocorrelation.
Explain the general difference that Helmus and Ives observed between spatial and non-spatial PDARs, and why this difference should be expected in both the species-area relationship (SAR) and the phylogenetic diversity-area relationship.

2.  Use Knitr to create a pdf of your completed PhyloTraits_handout.Rmd document, push it to GitHub, and create a pull request.
The due date for this assignment is February 25, 2015 at 12:00 PM (noon).
