---
title: "Phylogenetic Diversity - Communities"
author: "Z620: Quantitative Biodiversity, Indiana University"
date: "February 26, 2015"
header-includes:
   - \usepackage{array}
output: pdf_document
geometry: margin=2.54cm
---

## OVERVIEW
Taxonomic measures of $\alpha$- and $\beta$-diversity do not integrate evolutionary information into estimates of community diversity 
In this handout, we will introduce methods used in phylogenetic community ecology.
These methods will allow us to account for phylogenetic diversity and provide insight into the mechansisms that give rise to over- and under-dispersion (i.e., clustering) of biological communities. 

After completing this exercise you will know how to:

1.  measure phylogenetic $\alpha$ diversity
2.  measure phylogenetic $\beta$ diversity
3.  construct phylogenetic diversity-area relationships

## 1) SETUP
### A. Retrieve and Set Your Working Directory

```{r, results = 'hide'}
rm(list = ls())
getwd()
setwd("~/GitHub/QuantitativeBiodiversity/Assignments/PhyloCom")
```

### B. Load Packages 
We will be relying heavily on the R package `picante`. 
This package has many of the functions that are contained in the software Phylocom, which is used for the analysis of phylogenetic community structure and character evolution (http://phylodiversity.net/phylocom/).

After the initial installation of these packages using the `install.packages()` fucntion, let's load the packages and their dependencies with the `require()` function:

```{r, results = 'hide', message = FALSE, warning = FALSE} 
require("picante")
require("ape")
require("seqinr")
```

### C. Load Source Code
In addition to relying on the installation of contributed packages, this week we will also be using a source code file.
A source code file is a file containing user-defined functions that are required for an analysis.
The benefit of source files is that they can easily be used in multiple projects without fear of having code variation.
Here we will be using a source code file that includes a function for reading in the output files from the popular community sequencing software `mothur` (http://www.mothur.org/).

```{r}
source("./bin/MothurTools.R")
```

## 2) DESCRIPTION OF DATA
We will revisit the environmental, geographic, and biological data that we used a couple of weeks ago that was collected from the Brown County refuge ponds.
Perhaps a bit more info on the nature of the sequence data and how we generated the phylogeny. 
Also, maybe reference to the reference data base. 

## 3) LOADING OF DATA
```{r}
# Load Environmental Data
env <- read.table("data/20130801_PondDataMod.csv", sep = ",", header = TRUE)

# Load Site-by-Species Matrix
com <- read.otu(shared = "./data/INPonds.final.rdp.shared", cutoff = "1")

# Import Tree File and Create a `phylo` Object {ape}
ponds.cons <- read.alignment(file = "./data/INPonds.final.rdp.1.rep.fasta", format = "fasta")  
DNAbin <- as.DNAbin(ponds.cons) 
image.DNAbin(DNAbin, cex.lab = 0.50) 
seq.dist.jc <- dist.dna(DNAbin, model = "JC", pairwise.deletion = FALSE)
jc.tree <- bionj(seq.dist.jc)
```

# Examples from Picante Vignette (http://goo.gl/EYOZ55)
data(phylocom)
names(phylocom)

# mothur Generates a Newick formatted file so we'll probably need to use the `read.tree` function in {ape}
# We need to make sure that the names associate with the tree matach up with the names in the site-by-species matrix in `comm` below. 

phy <- phylocom$phy

# Community Data: Site-by-Species Matrix
comm <- phylocom$sample

# Traits Data: I don't think we'll using this, and least not for ponds
traits <- phylocom$traits

```{r}
# This is actually plotting samples onto the tree; labels (rownames) just correspond to "mock" samples, which demonstrate different patterns of dispersion; "include only species present in a community data set"
# Depending on number of OTUs in our samples, this could look pretty ugly. 

prunedphy <- prune.sample(comm, phy)

par(mfrow = c(2, 3))

for (i in row.names(comm)) {
  plot(prunedphy, show.tip.label = FALSE, main = i) 
  tiplabels(tip = which(prunedphy$tip.label %in% names(which(comm [i, ] > 0))), pch = 19, cex = 2)
  }
```

## 4) PHYLOGENETIC ALPHA DIVERSITY

### A.  Faith's Phylogenetic Diversity (PD)
In 1992, Daniel Faith developed a diversity metric called Faith's PD (http://goo.gl/wM08Oy).
A phylognetic tree containing the regional species pool is required. 
Then, the metric sums the branches lengths for each species in a sample from the root to the tip of the tree. 
Faith's PD can be implemented in R using the `pd` function in the picante package. 
The `pd()` function returns Faith's PD and species richness (SR), which is the number of species in a sample. 

```{r}
# Compares PD and Taxonomic Richness {picante}
pd <- pd(comm, phy, include.root=FALSE)
```

Let's compare the phylogenetic diversity estimates with taxonomic richness of our samples.
```{r}
# Biplot of Taxonomic Richness and Faith's PD
par(mar = c(5, 5, 2, 1) + 0.1)
plot(pd.result$SR, pd.result$PD, 
     pch = 20, col = "red", las = 1, asp = 1, xlim = c(0, 35), ylim = c(0, 35),
     xlab = "Taxonomic Richness", ylab = "Phylodiversity") 
abline(b = 1, a = 0, lty = 2)
text(20, 30, "1:1")

# Correlation of Taxonomic Richness and Faiths' PD
corr <- cor(pd.result$SR, pd.result$PD)
```

***Question 1***: Compare patterns of phylogenetic diversity and species richness. 
a.  Does taxonomic richness and phylodiversity correlate?
b.  Mathematically, why should they be correlated?  
c.  Under what conditions would you expect these two estimates of diversity to deviate from one another?

> ***Answer 1a***:  
> ***Answer 1b***:  
> ***Answer 1c***:  

We can address the statistical issues related to Faith's PD by conducting randomizations. 
The `ses.pd()` function in `picante` allows one to compare the observed value for Faith's PD against randomized data derived from different null models. 

```{r}
# Estimate Standardized Effect Size of PD via Randomization {picante} 
ses.pd <- ses.pd(comm, phy, null.model = "taxa.labels", runs = 1000, iterations = 999)
```

***Question 2***: Using the `help()` function, identify two null models that can be used with `ses.pd()`
Compare and interpret the output from these randomizations. 

> ***Answer 2***:  

### B.  Net Relatedness Index (NRI)
One common way to test for phylogenetic clustering and overdispersion is to use the Net Relatedness Index (NRI) 
The first step to NRI is to calculate the mean phylogenetic distance (MPD), which is the sum of phylogenetic pairwise distance among taxa. 
NRI is often estimated as the observed MPD minus the mean MPD generated via randomization divided by the standard deviation of the randomized MPD. 
The resulting value is multiplied by -1 so that negative values correspond with overdispersion and positive values correspond with clustering. 
Let's calucate NRI:

```{r}
# Create a Phylogenetic Distance Matrix {picante}
phydist <- cophenetic(phy)

# Estimate Standardized Effect Size of NRI via Randomization {picante} 
ses.mpd <- ses.mpd(comm, phydist, null.model = "taxa.labels", abundance.weighted = FALSE, runs = 1000)
 
# Calculate NRI
NRI <- as.matrix(-1 * ((ses.mpd[,1] - ses.mpd[,2]) / ses.mpd[,3]))
rownames(NRI) <- row.names(ses.mpd)
colnames(NRI) <- "NRI"
```

### C.  Nearest Taxon Index (NTI)
Another way to test for phylogenetic clustering and oversdispersion in a sample is to use the Nearest Taxon Index (NRI).
This index is similar to NRI, but uses the mean nearest phylogenetic neighbor distance (MNND) insted of MPD. 
Just like NRI, we peform randomizations and use this information to estimate that standardized effect size. 
Last, we multiply by -1 so that negative values correspond with overdispersion and positive values correspond with clustering. 

```{r}
# Estimate Standardized Effect Size of NRI via Randomization {picante} 
ses.mntd <- ses.mntd(comm, phydist, null.model = "taxa.labels", abundance.weighted = FALSE, runs = 1000)

# Calculate NTI
NTI <- as.matrix(-1 * ((ses.mntd[,1] - ses.mntd[,2]) / ses.mntd[,3]))
rownames(NTI) <- row.names(ses.mntd)
colnames(NTI) <- "NTI"
```

***Question 3***: In the NRI and NTI examples above, the arguments "abundance.weighted = FALSE" means that the indices were calculated on presence-absence data.
Modify and rerun the code so that NRI and NTI are caculated using abundance data. 
How does this affect the interpretation of NRI and NTI?

> ***Answer 3***:  


## 5) PHYLOGENETIC BETA DIVERSITY 

```{r}
comdist.result <- comdist(comm, phydist)
comdist.result
comdist.clusters <- hclust(comdist.result)
plot(comdist.clusters)

#  phyloordination ?Phylogenetic beta diversity measures can be used with any method based onvmeasuring among-community distances. For example, they could be used in a cluster analysis or phyloordination to group communities based on their evolutionaryvsimilarity, or they could be compared with spatial or environmental distances separating communities using a Mantel test. The code below calculates MPD betweenvpairs of communities, and uses these phylogenetic distances to cluster communities based on their phylogenetic similarity:
```

Ordinations: PCoA taxo  vs. UNIFRAC
http://www.inside-r.org/packages/cran/picante/docs/unifrac
http://cran.r-project.org/web/packages/GUniFrac/GUniFrac.pdf
http://joey711.github.io/phyloseq-demo/unifrac.html  <- out of date. not worth it

```{r}

unifrac(comm, phydist)
GUniFrac(comm, phydist)   # Tree Must be rooted


GUniFrac(otu.tab, tree, alpha = c(0, 0.5, 1))

# Calculate the UniFracs
unifracs <- GUniFrac(otu.tab.rff, throat.tree, alpha=c(0, 0.5, 1))$unifracs

#  Combine unweighted and weighted UniFrac for testing
PermanovaG(unifracs[, , c("d_1", "d_UW")] ~ groups)
# Combine d(0), d(0.5), d(1) for testing
PermanovaG(unifracs[, , c("d_0", "d_0.5", "d_1")] ~ groups)



phylosor.rnd(samp,tree, cstSor=TRUE, null.model=c("taxa.labels",
    "frequency","richness","independentswap","trialswap"),
    runs=999, iterations=1000)

```


## 6) PHYLOGENETIC DIVERSITY-AREA RELATIONSHIP



## 7) HOMEWORK 

1.  
2.  Use Knitr to create a pdf of your completed PhyloTraits_handout.Rmd document, push it to GitHub, and create a pull request.
The due date for this assignment is February 25, 2015 at 12:00 PM (noon).
