---
title: "Phylogenetic Diversity - Communities"
author: "Z620: Quantitative Biodiversity, Indiana University"
date: "February 26, 2015"
header-includes:
   - \usepackage{array}
output: pdf_document
geometry: margin=2.54cm
---

## OVERVIEW
Taxonomic measures of $\alpha$- and $\beta$-diversity do not integrate evolutionary information into estimates of community diversity 
In this handout, we will introduce methods used in phylogenetic community ecology.
These methods will allow us to account for phylogenetic diversity and provide insight into the mechansisms that give rise to over- and under-dispersion (i.e., clustering) of biological communities. 

After completing this exercise you will know how to:

1.  measure phylogenetic $\alpha$ diversity
2.  measure phylogenetic $\beta$ diversity
3.  construct phylogenetic diversity-area relationships

## 1) SETUP
### A. Retrieve and Set Your Working Directory

```{r, results = 'hide'}
rm(list = ls())
getwd()
setwd("~/GitHub/QuantitativeBiodiversity/Assignments/PhyloCom")
```

### B. Load Packages 
We will be relying heavily on the R package `picante`. 
This package has many of the functions that are contained in the software Phylocom, which is used for the analysis of phylogenetic community structure and character evolution (http://phylodiversity.net/phylocom/).

After the initial installation of these packages using the `install.packages()` fucntion, let's load the packages and their dependencies with the `require()` function:

```{r, results = 'hide', message = FALSE, warning = FALSE} 
require("picante")
require("ape")
require("seqinr")
require("reshape")
```

### C. Load Source Code
In addition to relying on the installation of contributed packages, this week we will also be using a source code file.
A source code file is a file containing user-defined functions that are required for an analysis.
The benefit of source files is that they can easily be used in multiple projects without fear of having code variation.
Here we will be using a source code file that includes a function for reading in the output files from the popular community sequencing software `mothur` (http://www.mothur.org/).

```{r}
source("./bin/MothurTools.R")
```

## 2) DESCRIPTION OF DATA
We will revisit the environmental, geographic, and biological data that we used a couple of weeks ago that was collected from the Brown County refuge ponds.
Perhaps a bit more info on the nature of the sequence data and how we generated the phylogeny. 
Also, maybe reference to the reference data base. 

## 3) LOADING OF DATA
First, we'll load the environmental data, along with the site-by-species matrix:

```{r}
# Load Environmental Data
env <- read.table("data/20130801_PondDataMod.csv", sep = ",", header = TRUE)

# Need to remove following 7 samples from env and comm
# BC010, HNF191, HNF221, HNF236, HNF237, YSF46, YSF67

# Load Site-by-Species Matrix
comm <- read.otu(shared = "./data/INPonds.final.rdp.shared", cutoff = "1")

# Select Only DNA Using GREP command and rename using gsub
comm <- comm[grep("*-DNA", rownames(comm)), ] 
rownames(comm) <- gsub("\\-DNA", "", rownames(comm)) 
rownames(comm) <- gsub("\\_", "", rownames(comm))

# Remove Zero-Occurrence Taxa
comm <- comm[ , colSums(comm) > 0]

# Import Taxonomy Data using read.tax from source code
tax <- read.tax(taxonomy = "./data/INPonds.final.rdp.1.cons.taxonomy")
```

Now, let's load and process some of the phylogenetic data:

```{r}
# Import the Alignment File {seqinr}
ponds.cons <- read.alignment(file = "./data/INPonds.final.rdp.1.rep.fasta", format = "fasta")  

# Rename OTUs in FASTA File
ponds.cons$nam <- gsub("\\|.*$", "", gsub("^.*?\t", "", ponds.cons$nam))

# Import Outgroup FASTA File {seqinr}
outgroup <- read.alignment(file = "./data/methanosarcina.fasta", format = "fasta")

# Convert Alignment File to DNAbin Object {ape}
DNAbin <- rbind(as.DNAbin(outgroup), as.DNAbin(ponds.cons))

# Visusalize Sequence Alignment {ape}
image.DNAbin(DNAbin, show.labels=T, cex.lab = 0.05, las=1) 

# Create Distance Matrix with the Jukes Cantor "JC" Model {ape}
seq.dist.jc <- dist.dna(DNAbin, model = "JC", pairwise.deletion = FALSE)

# Use Neighbor Joining Algorithm to Construct a Full Tree (DNA and RNA sequences) {ape}
phy.all <- bionj(seq.dist.jc)

# Drop Tips of Zero-Occurrence OTUs (Removes Taxa Only Found via RNA sequences) {ape}
phy <- drop.tip(phy.all, phy.all$tip.label[!phy.all$tip.label %in% 
                                             c(colnames(comm), "Methanosarcina")])

# Identify Outgroup Sequence
outgroup <- match("Methanosarcina", phy$tip.label)

# Root the Tree {ape}
phy.rooted <- root(phy, outgroup, resolve.root = TRUE)

# Plot the Rooted Tree {ape}
par(mar = c(1,1,2,1) + 0.1)
plot.phylo(phy.rooted, main = "Neigbor Joining Tree", "phylogram", 
           use.edge.length = FALSE, direction = "right", cex = 0.6, label.offset = 1)

# Save tree as phy
phy <- phy.rooted
```

## Example from Picante Vignette: way to move forward while figuring out the pond data set

```{r}
# From Picante Vignette (http://goo.gl/EYOZ55)
data(phylocom)
names(phylocom)

# We need to make sure that the names associate with the tree match up with the names in the site-by-species matrix in `comm` below. 
phy.eg <- phylocom$phy

# Community Data: Site-by-Species Matrix
comm.eg <- phylocom$sample
```

## 4) PHYLOGENETIC ALPHA DIVERSITY

### A.  Faith's Phylogenetic Diversity (PD)
In 1992, Daniel Faith developed a diversity metric called Faith's PD (http://goo.gl/wM08Oy).
The metric sums the branch lengths for each species in a sample from the root to the tip of the tree.
The value of the metric captures the evolutionary history of an assemblage. 
Higher PD values indicate that an assemblage contains more evoluiotnarily divergent taxa, while lower PD values indicate that an assemblage contains taxa with a more restricted evolutionary history. 

Faith's PD can be implemented in R using the `pd()` function in the `picante` package. 
A phylognetic tree containing the regional species pool is required. 
In addition to returning Faith's PD, the `pd()` function also returns species richness (SR).
SR is the same as observed richness ($\S_{obs}), which we covered in the $\alpha$ diversity module. 

```{r}
# Compares PD and Taxonomic Richness {picante}
pd <- pd(comm, phy, include.root = FALSE)
```

Let's compare the phylogenetic diversity (PD) estimates with taxonomic richness (SR) of our samples.
We'll use a log-log transformation, so that the slope of the relationship gives us a power-law exponent which describes how PD scales with SR.

```{r}
# Biplot of SR and PD, log-transforming
par(mar = c(5, 5, 4, 1) + 0.1)

plot(log(pd$SR), log(pd$PD), 
     pch = 20, col = "red", las = 1,

     xlab = "log(SR)", ylab = "log(PD)", cex.main = 0.85,
     main="Phylodiversity (PD) vs. Taxonomic richness (SR)")


OLS <- lm('log(pd$PD) ~ log(pd$SR)')
abline(OLS, col = "red", lw = 2)
exponent <- round(coefficients(OLS)[2], 3)
legend("topleft", legend=paste("Scaling exponent = ", exponent, sep = ""), 
       bty="n", lw=2, col = "red") 
```

***Question 1***: Compare patterns of PD and SR. 
a.  Describe the relationship between taxonomic richness and phylodiversity?
b.  Mathematically, why should they be correlated?  
c.  Under what conditions would you expect these two estimates of diversity to deviate from one another?

> ***Answer 1a***:  
> ***Answer 1b***:  
> ***Answer 1c***:  

### B. Randomizations and Null Models
Randomizations are a way of resampling our data to assess whether or not observed patterns are different from a null expectation. 
A number of the functions in `picante` allow us to specify different null models as an argument. 
These null models can control for features such as species richness, species occurrence frequencey, and the diversity of the regional species pool. 
The following table describes some of the null models that are available to us when using `picante`:

\begin{center}
\hyphenpenalty 10000
\exhyphenpenalty 10000
\begin{tabular}{ m{5cm}  m{10cm} }
  \textbf{Null Model} & \textbf{Description} \\
  \hline \hline \\ [-1.5ex]
  \textbf{taxa.labels} & 
  Shuffles taxa labels across tips of phylogeny (across all taxa included in phylogeny) \\
  \\ [-1.5ex]
  \textbf{richness} & 
  Randomizes community data matrix abundances within samples (maintains sample species richness) \\
  \\ [-1.5ex]  
  \textbf{frequency} & 
  Randomizes community data matrix abundances within species (maintains species occurence frequency) \\
  \\ [-1.5ex]   
  \textbf{sample.pool} & 
  Randomizes community data matrix by drawing species from pool of species occurring in at least one community (sample pool) with equal probability \\
  \\ [-1.5ex]  
    \textbf{phylogeny.pool} & 
  Randomize community data matrix by drawing species from pool of species occurring in at least one community (sample pool) with equal probability \\
  \\ [-1.5ex]
  \textbf{independentswap} & 
  Randomizes community data matrix with the independent swap algorithm (Gotelli 2000) maintaining species occurrence frequency and sample species richness \\
  \\ [-1.5ex]
  \textbf{trialswap} & 
  Randomizes community data matrix with the trial-swap algorithm (Miklos \& Podani 2004) maintaining species occurrence frequency and sample species richness \\
  \\ [-1.5ex]
    \hline
\end{tabular}
\end{center}

Now, we are going to use the `ses.pd()` function in `picante`. 

This function estimates the standardized effect size ("ses") by comparing the observed PD value to the PD values generated from a tree randomized under a null model (see table above). 

```{r}
# Estimate Standardized Effect Size of PD via Randomization {picante} 
ses.pd <- ses.pd(comm, phy, null.model = "taxa.labels", runs = 100, include.root = FALSE)
```

***Question 2***: Using the `help()` function and the table above, identify two null models that can be used with `ses.pd()`.
Rerun `ses.pd()` with these null models.
Compare and interpret the output. 

> ***Answer 2***:  

### B.  Net Relatedness Index (NRI)
One common way to test for phylogenetic clustering and overdispersion is to use the Net Relatedness Index (NRI).
<<<<<<< HEAD
NRI is caculated with the following equation: $\ - \frac{MPD_{obs} - \overline{MPD_{null}}}{\sigma \left (MPD_{null} \right )}$ where $MPD_{obs}$ is the mean phylogenetic distance cacluated from pairwise branch lengths in a sample, $\overline{MPD_{null}}$ is the mean of the MPD values generated from a tree randomized under a null model, and $\sigma MPD_{null}$ is the standard deviation of the MPD values generated from a tree randomized under a null model (see table above). 
 
Negative NRI values indicate that a sample is phylogenetically overdispersed; that is, taxa are less related to one another than expected under the null model.
Positive NRI values indicate that a sample is phylogentically underdispersed, or clustered, such that taxa are more closely related to one another than expected under the null model. 

Let's use `picante` to calculate NRI.
The code below will generate a phylogenetic distance matrix, which is basically a variance-covariance matrix similar to the resemblance matrices we have used in previous lessons.
In the variance-covariance matrix, the off-diagonal terms are covariances, i.e., sum((Xi - Xm)(Yi - Ym)).
Since the covariance of a variable with itself is the variance, the diagonal terms are variances: sum((Xi - Xm)*(Xi - Xm)).
So, once more we see that a matrix (e.g. site-by-species matrix, resemblance matrix, phylogenetic distance matrix) has become a primary data structure in our analysis of biodiversity.

```{r}
# Create a Phylogenetic Distance Matrix {picante}
phydist <- cophenetic.phylo(phy)


# Estimate Standardized Effect Size of NRI via Randomization {picante} 
ses.mpd <- ses.mpd(comm, phydist, null.model = "taxa.labels", abundance.weighted = FALSE, runs = 100) # 12:10 ...what's this for? --Ken

# Calculate NRI
NRI <- as.matrix(-1 * ((ses.mpd[,2] - ses.mpd[,3]) / ses.mpd[,4]))
rownames(NRI) <- row.names(ses.mpd)
colnames(NRI) <- "NRI"
```

### It seems like these NRI values are all really really negative. Is that right? {MEM}

### C.  Nearest Taxon Index (NTI)
Another way to test for phylogenetic clustering and oversdispersion in a sample is to use the Nearest Taxon Index (NRI).
This index is similar to NRI, but uses the mean nearest phylogenetic neighbor distance (MNND) insted of MPD. 
Just like NRI, we peform randomizations and use this information to estimate that standardized effect size. 
Last, we multiply by -1 so that negative values correspond with overdispersion and positive values correspond with clustering. 

```{r}
# Estimate Standardized Effect Size of NRI via Randomization {picante} 
ses.mntd <- ses.mntd(comm, phydist, null.model = "taxa.labels", abundance.weighted = FALSE, runs = 100)

# Calculate NTI
NTI <- as.matrix(-1 * ((ses.mntd[,2] - ses.mntd[,3]) / ses.mntd[,4]))
rownames(NTI) <- row.names(ses.mntd)
colnames(NTI) <- "NTI"
```

***Question 3***: In the NRI and NTI examples above, the arguments "abundance.weighted = FALSE" means that the indices were calculated on presence-absence data.
Modify and rerun the code so that NRI and NTI are caculated using abundance data. 
How does this affect the interpretation of NRI and NTI?

> ***Answer 3***:  


## POTENTIAL EXERCISE: HAVE STUDENTS LOOK FOR CORRELATIONS BETWEEN NRI/NTI AND ENVIRONMENTAL VARIABLES. 
## DOES CLUSTERING/DISPERSION CHANGE ALONG GRADIENTS OR WITH PD? IF SO, WHAT ARE SOME DIFFERENT EXPLANATIONS?

```{r}
disp <- data.frame(NRI, NTI)
```


## 5) PHYLOGENETIC BETA DIVERSITY 

```{r}
comdist.result <- comdist(comm, phydist)
comdist.result
comdist.clusters <- hclust(comdist.result)
plot(comdist.clusters)

#  phyloordination ?Phylogenetic beta diversity measures can be used with any method based onvmeasuring among-community distances. For example, they could be used in a cluster analysis or phyloordination to group communities based on their evolutionaryvsimilarity, or they could be compared with spatial or environmental distances separating communities using a Mantel test. The code below calculates MPD between pairs of communities, and uses these phylogenetic distances to cluster communities based on their phylogenetic similarity:
```

Ordinations: PCoA taxo  vs. UNIFRAC
http://www.inside-r.org/packages/cran/picante/docs/unifrac
http://cran.r-project.org/web/packages/GUniFrac/GUniFrac.pdf
http://joey711.github.io/phyloseq-demo/unifrac.html  <- out of date. not worth it

```{r}

unifrac(comm, phydist)
# GUniFrac(comm, phydist)   # Tree Must be rooted


GUniFrac(otu.tab, tree, alpha = c(0, 0.5, 1))

# Calculate the UniFracs
unifracs <- GUniFrac(otu.tab.rff, throat.tree, alpha=c(0, 0.5, 1))$unifracs

#  Combine unweighted and weighted UniFrac for testing
PermanovaG(unifracs[, , c("d_1", "d_UW")] ~ groups)
# Combine d(0), d(0.5), d(1) for testing
PermanovaG(unifracs[, , c("d_0", "d_0.5", "d_1")] ~ groups)



phylosor.rnd(samp,tree, cstSor=TRUE, null.model=c("taxa.labels",
    "frequency","richness","independentswap","trialswap"),
    runs=999, iterations=1000)

```


## 6) PHYLOGENETIC DIVERSITY-AREA RELATIONSHIP (PDAR)


```{r, echo=FALSE}
# Ken's SAR function
SAR <- function(OTUs){
  Alist <- c()
  Slist <- c()
  
  num.ponds <- c(1,2,4,8,16,32,51)
  for (i in num.ponds) {   
    areas <- c() # hold iterated area values 
    Ss <- c() # hold iterated S values
    
    for(j in 1:100){
      pond.sample <- sample(51, replace = FALSE, size = i) 
      area <- 0
      cum.abs <- vector(mode="numeric", length=length(comm))
  
      for (k in pond.sample) { # Loop through each randomly drawn pond
        area <- area + pond.areas[k] # aggregating area, (incrementing)
        cum.abs <- cum.abs + as.numeric(comm[k, ])
    
      } # End random pond samples loop
      S <- length(cum.abs[cum.abs > 0])
      areas <- c(areas, area)
      Ss <- c(Ss, S)
    }
    Alist <- rbind(Alist, mean(areas))
    Slist <- rbind(Slist, mean(Ss))
    print(c(mean(areas), mean(Ss)))
  }
  return(cbind(log(Alist), log(Slist)))
}
```

Generate the SAR (random aggregation)

```{r, echo=FALSE}
pond.areas <- as.vector(pi * (env$Diameter/2)^2) # Find areas of all 51 ponds
sar <- SAR(comm)
sar <- as.data.frame(sar)

plot(sar, xlab="log(Area)", ylab="log(Richness)", 
     main="Species-area relationship", col="red")

OLS <- lm('sar$V2 ~ sar$V1')
abline(OLS, col = "red", lw = 2)

slope <- round(coefficients(OLS)[2], 3)
legend("topleft", legend = paste("slope =", slope), bty = "n", lw = 2, col = "red")  
```


Helmus and Ives (2012) studied how phylogenetic diversity changes with increasing area.
This relationship, abbreviated PDAR, is analogous to the species-area relationship (SAR) we learned about in Geographical Ecology.
Remember that the SAR is a cumulative relationship (S = cA^z) and so, cannot be negative.
In contrast the PDAR can increase or decrease with area, which provoked Helmus and Ives (2012) to develop theoretical expectations for the effects of forces or processes on the shape of the PDAR.

Helmus and Ives use the Phylogenetic species variability (PSV) metric to quantify phylogenetic diversity.
PSV quantifies how phylogenetic relatedness decreases the variance of a hypothetical neutral trait shared by all species in a community. 
Type 'help(psv)' to learn more.

In their study, Helmus and Ives, increased area by aggregating plots at random either with or without respect to whether the plots were adjacent.
Both sampling methods produce roughly the same result.
Here, we will construct PDARs using the random aggregation approach of Helmus and Ives (2012).
**Fortunately, Helmus and Ives (2012) did this in a way that is similar to how we constructed SARs in Geographical Ecology.**

Let's begin by writing a function to generate the PDAR.

```{r}
pdac.psv <- function(comm, tree, PD.index='PSV'){
  # Use the PSV index of Helmus and Ives (2012) by default
  # Using PD.index, we've left room to call other diversity metrics
  
  areas <- c() # an object to hold areas
  diversity <- c() # an object hold diversity
  
  num.plots <- c(2,4,8,16,24,48) # increase numbers of plots by 2, roughly doubling area
  
  for (i in num.plots){    
    areas.iter <- c() # hold areas from iterations; we'll find the mean from this
    diversity.iter <- c() # hold estimated diversity from iterations; we'll find the mean
    
    for (j in 1:10){ # Iterate 100 times for each sample size
      pond.sample <- sample(51, replace = FALSE, size = i) # sampling w/out replacement
      area <- 0 # a variable to hold accumulating area
      sites <- c() # a vector to hold accumulating taxa
      
      for (k in pond.sample) { # Loop through each randomly drawn pond...  
        area <- area + pond.areas[k] # aggregating area...
        sites <- rbind(sites, comm[k, ]) # and sites
        }
      
      areas.iter <- c(areas.iter, area) # concatenate the area to areas.iter
      if (PD.index == 'PSV'){ # Other 'if' statements could call other metrics        
        psv.vals <- psv(sites, tree, compute.var=FALSE) # calculate PSV
        psv <- psv.vals$PSVs[1] 
        diversity.iter <- c(diversity.iter, as.numeric(psv))
        }

      }
      diversity <- c(diversity, mean(diversity.iter)) # let diversity be the mean PSV
      areas <- c(areas, mean(areas.iter)) # let area be the average area
      print(c(i, mean(diversity.iter), mean(areas.iter))) # print as we go
    }
  return(cbind(areas, diversity)) # return vectors of areas (x) and diversity (y)
  }
```


Now, let's examine the PDAR for our pond data set.
We will investigate the relationship of phylogenetic diversity to area using both Spearman's correlation coefficient (S) and Pearson's correlation coefficient (P).
It is informative to use both because while S is computed on ranks and depicts monotonic relationships (the degree to which the relationship is continually increasing or decreasing), P is computed on the observed values and therefore depicts linear relationships.
For example, if you S > P then the correlation is more monotonic than linear and vice versa.

```{r}
pdac <- pdac.psv(comm, phy.rooted) # Compute the PDAR
pdac <- as.data.frame(pdac)

Pearson <- cor.test(pdac$areas, pdac$diversity, method="pearson")
P <- round(Pearson$estimate,2)
Pp <- round(Pearson$p.value,3)

Spearman <- cor.test(pdac$areas, pdac$diversity, method="spearman")
S <- round(Spearman$estimate,2)
Sp <- round(Spearman$p.value,3)

plot(pdac, xlab='Area', ylab='PSV', main='Phylogenetic diversity-area relationship', col='red')
legend("topleft", legend=paste("Spearman correlation =", S, "p =", Sp,
                               "\nPearson correlation =", P, "p =", Pp), bty="n", lw=0, col = "red") 
```

## 6) PHYLOGENETIC DISTANCE-DECAY RELATIONSHIP (PDD)
Remember from the Geographical Ecology lesson that the Distance Decay relationship reflects the spatial autocorrelation of community similarity.
That is, geographically near communities should be more similar than geographically distant communities.
Given that similarity in community composition generally decreases with geographic distance, let's apply the concept of spatial autocorrelation to phylogenetic diversity and ask whether similarity in phylogenetic diversity also decreases with distance.
Once more, we will use the PSV metric of Helmus and Ives (2012).

To generate the PDD, we will need to calculate geographic distances based on lat-long coordinates.
Let's use the `fossil` package for this.
```{r}
require("fossil")
```

Next, let's write a function to find the geographic distances and the differences in phylogenetic diversity between all unique pair-wise combinations of ponds.

```{r}

PDD <- function( ){

  pond.pairs <- as.data.frame(combn(c(1:51), 2)) # find all unique pair-wise combinations
  length(pond.pairs) # The number of pairs is simply (51*50)/2

  dists <- c() # an object to hold geographic distances
  diffs <- c() # an object to hold phylogenetic differences
  
  for (pair in pond.pairs){
    print(pair)
    p1 <- pair[1]
    p2 <- pair[2]
    
    # Find phylogenetic difference
    psv1 <- psv(comm[p1, ], phy, compute.var=FALSE)[1] # PSV for the first pond  
    psv2 <- psv(comm[p2, ], phy, compute.var=FALSE)[1] # PSV for the second pond    
    diffs <- rbind(diffs, abs(psv1 - psv2))
    
    # Find geographic distance
    lat.long <- rbind(c(env$long[p1], env$lat[p1]), c(env$long[p2], env$lat[p2]))
    geo.dist <- earth.dist(lat.long, dist=TRUE) # in kilometers
    dists <- rbind(dists, as.numeric(geo.dist))
  }
  
  return(cbind(dists, diffs))
}
```


Now, let's generate and analyze the distance decay relationship for phylogenetic diversity.
We'll be using Helmus and Ives (2012) metric, i.e., phylogenetic species variability (PSV).

```{r}
pdd <- PDD() # Generate the phylogenetic diversity distance decay
pdd2 <- as.data.frame(pdd)

Pearson <- cor.test(pdd2$V1, pdd2$PSVs, method="pearson")
P <- round(Pearson$estimate,2) # Pearson's
Pp <- round(Pearson$p.value,3) # p-value

Spearman <- cor.test(pdd2$V1, pdd2$PSVs, method="spearman")
S <- round(Spearman$estimate,2) # Spearman's
Sp <- round(Spearman$p.value,3) # p-value

plot(pdd2, xlab='Distance (Km)', ylab='Difference in phylodiversity, PSV',
     main='Difference in phylogenetic diversity
     with increasing distance lags', col='red')

legend("topleft", legend=paste("Spearman correlation =", S, "p =", Sp,
        "\nPearson correlation =", P, "p =", Pp), bty="n", lw=0, col = "red") 
```

## 7) HOMEWORK 

1.  In their study of phylogenetic diversity-area relationships (PDARs), Helmus and Ives, increased area by aggregating plots in two ways.
First, they aggregated plots with respect to whether the plots were adjacent; what they called 'spatial'.
Second, they aggregated plots at random ('non-spatial').
While both spatial and non-spatial sampling methods capture the effect of increasing area, sampling with respect to location (i.e. spatial) also captures the effect of increasing distance or spatial autocorrelation.
Explain the general difference that Helmus and Ives observed between spatial and non-spatial PDARs, and why this difference should be expected in both the species-area relationship (SAR) and the phylogenetic diversity-area relationship.

2.  Use Knitr to create a pdf of your completed PhyloTraits_handout.Rmd document, push it to GitHub, and create a pull request.
The due date for this assignment is February 25, 2015 at 12:00 PM (noon).
