---
title: "Phylogenetic Diversity - Traits"
author: "Z620: Quantitative Biodiversity, Indiana University"
date: "February 20, 2015"
header-includes:
   - \usepackage{array}
output: pdf_document
geometry: margin=2.54cm
---

## OVERVIEW
Up to this point, we have been focusing on patterns taxonomic diversity in Quantitative Biodiversity. 
Although taxonomic diversity is an important dimension of biodiversity, it does not consider the evolutionary history and relatedness of species. 
The goal of this exercise is to introduce basic concepts of phylogenetic diversity. 

After completing this exercise you will know how to:

1.  create phylogenetic trees
2.  map traits onto phylogenetic trees
3.  test for phylogenetic signal of traits on a phylogenetic tree
4.  test for relationships among traits using phylogenetic independent contrasts 

## 1) SETUP
### A. Retrieve and Set Your Working Directory

```{r, results = 'hide'}
rm(list = ls())
getwd()
setwd("~/GitHub/QuantitativeBiodiversity/Assignments/PhyloTraits")
```

### B. Load Packages 
There are many contributed R packages that have been developed for conducting phylogenetic analyses (http://goo.gl/DtU16j). 
We will rely heavily on the `ape` (http://goo.gl/nExs0) and `phangorn` (http://goo.gl/iWU6Vt) packages, which peform a wide range of phylogenetic analyses. 
We will supplement this with other packages for some specialized tasks such as sequence alignment (`muscle`), reading of alinged sequences (`seqinr`), and testing for phylogenetic signal (`geiger`).
After the initial installation of these packages using the `install.packages()` fucntion, let's load the packages and their dependencies with the `require()` function:

```{r, results = 'hide', message = FALSE, warning = FALSE} 
require("muscle")
require("ape")
require("seqinr")
require("phangorn")
require("geiger") 
```

## 2) DESCRIPTION OF DATA
The maintenance of biodiversity is thought to be influenced by **trade-offs** among species in certain functional traits. 
One such trade-off involves the ability of a highly specialized species to peform exceptionally well on a particular resource. 
This strategy may be advantageous if the specialist lives in an environment where its preferred resource is found in a relatively high and constant supply. 
However, the specialist strategy may not be optimal if the species finds itself itself in an environment where resource conditions are more variable.
Under these conditions, it may be beneficial if a species is capable of using a wider range of resources, even though there may be a fitness cost associated with being a generalist. 
We set out to test this hypothesized trade-off in bacteria by assessing generalist and specialist strategies on different forms of phopshorus, a resource which commonly limits growth and reproduction of organisms inhabiting freshwater ecosystems.
We collected 39 strains of bacteria from two lakes. 
After identifying these bacteria based on their 16S rRNA gene sequences, we measured their growth on 18 different forms of phosphorus. 
We will take a phylogenetic approach to mapping phosphorus resource use onto a phylogenetic tree while testing for specialist-generalist trade-offs. 

## 3) SEQUENCE ALIGNMENT
After generating or retrieving sequence data, the first step of a phylogenetic analysis is to peform an alignment. 
This can be done with nucleotides (DNA or RNA) or proteins (amino acids). 
There are various methods for aligning sequences, but the basic premise is to use an algorithm to arrange sequences relative to one another based on conserved regions so that downstream phylogenetic analyses can be performed.
In R, there are at least two packages that can be used for aligning sequences: `ape` and `muscle`. 
In `ape`, alignments can be done with the functions `clustal`, `muscle`, and `tcoffee`, all which have a range of options. 
However, these functions require that other programs be installed on your computer.
Therefore, for this exericse, we are going to use the `muscle` package in R becasue it will align sequences in a "stand alone" fashion without the need for installing additional software. 

### A.  Examining a FASTA-File 
We will read in a **FASTA-formatted** file into the `muslce` alignment function.
This is a very common text-based format that is used across platforms and disciplines. 
Let's take a second to look at the *p.isolates.fasta* file. 
Open the terminal and change directory to the data folder. 
You can then type "nano p.isolates.fasta" at the command line. 
This will open the FASTA file in nano, which is a text editing program. 
You can use Ctrl+V and Ctrl+Y to scroll towards the top and bottom of the file, respectively.
When you are done, you can close the file by hitting Ctrl+X.
Do not save any inadvertent changes to *p.isolates.fasta*.

### B.  Performing an Alignment 
Now, let's align our sequences. 
Using the `muscle` fuction, we will read in our FASTA file and create a **FASTA-aligned* output file (with an .afa extension).  

```{r}
muscle::muscle(seqs="./data/p.isolates.fasta", out = "./data/p.isolates.afa") 
```

***Question 1***: Using nano or your favorite text editor, compare the `p.isolates.fasta` file and the `p.isolates.afa` file. 
Describe the differences that you observe between the two files. 

> ***Answer 1***: 

### C.  Visualizing the Alignment
It is good practice to view your sequences after alignment. 
This may not be practical for large databases, but in our example, the alignment only contains sequences for 39 bacterial species plus a distantly related reference sequence that we will use as an **outgroup**. 
To visualize our alingment, first, we will read in the FASTA-aligned file using the `read.alignment` function in the `seqinr` package. 
Then, using the `ape` package, we will convert the alignnment file it into a **DNAbin object**, which is a bit-level coding scheme that allows R to store and manipulate sequence data.
Last, we will visualize the alignment by color-coding nucelobases different colors, where red = adenine, yellow = guanine, green = cytosine, blue = thymidine, grey = ambiguous call, and black = alignment gaps. 

```{r}
# Read Alignment File {seqinr}
read.aln <- read.alignment(file = "./data/p.isolates.afa", format = "fasta")  

# Convert Alignment File to DNAbin Object {ape}
p.DNAbin <- as.DNAbin(read.aln) 

# Identify Base Pair Region of 16S rRNA Gene to Visuzlize
window <- p.DNAbin[,100:500] 

# Command to Visusalize Sequence Alignment {ape}
image.DNAbin(window, cex.lab = 0.6) 

# Optional Code Adds Grid to Help Visualize Rows of Sequences 
grid(ncol(window), nrow(window), col = "lightgrey") 
```

***Question 2***:  Make some observations about the `muscle` alingmnet of the 16S rRNA gene sequences for our bacterial isolates and the outgroup, *Methanosarcina*, which belongs to the archaeal domain of life. 
Move along the alignment by changing the values in the `window` object.  
a.  Approximatley how long were our reads?  
b.  What regions do you think would be most appropriate for phylogenetic inference?  

> ***Answer 2a***:   
> ***Answer 2b***:  

## 4) MAKING A PHYLOGENETIC TREE
Once you have aligned your sequences, the next step is to construct a phylogenetic tree.
Not only is this effective for visualizing the evolutionary relationship among taxa, but as you will see later, the information that goes into a phylogenetic tree is needed for downstream hypothesis testing. 

There are a number of ways to construct a phylogenetic tree. 
For example, **maximum parsimony** is character-based approach that involves identifying one of many possible trees based upon the minimum number of evolutionary "steps"" (or mutations). 
We are not going to cover this method, but it is good to know that maximum parsimony can easily be peformed in R using the `phangorn` package. 
Instead, we wil empahasize distance-based approaches that are used for generating trees, specifically **neighbor-joining**. 
Neighbor joining is a relatively fast algorithm and often generates correct tree topologies.
Moreover, neighbor joining can serve as a starting point for assessing different evolutionary models using **maximum likelihood** methods, which we will explore in this handout. 
Last, phylogenetic trees can be constructed using **Bayesian** methods in the R package `phangorn`.
Although Bayesian phylogenetic methods are used by a lot of researchers, we do not have time to cover these approaches in Quantiative Biodiversity. 

##MARIO: CAN YOU START ADDING TEXT TO FILL OUT THIS TABLE?
\begin{center}
\hyphenpenalty 10000
\exhyphenpenalty 10000
\begin{tabular}{ m{5cm} m{10cm} }
  \textbf{Tree Method} & \textbf{Properties} \\
  \hline \hline \\ [-1.5ex]
  \textbf{UPGMA} & 
  A simple agglomerative hierarchical clustering technique. Assumes a constant rate of evolution. Likely not appropriate for infering phylogenetic relationships \\
  \\ [-1.5ex]
  \textbf{Neighbor Joining} & 
  An agglomerative hierarchical clustering technique. Based on a user defined distance matrix. Advantage: Fast  \\
  \\ [-1.5ex]  
  \textbf{Maximum Parsimony} & 
  A non-parametric statistical method for constructing phylogenies. The final tree is the one that uses the least evolutionary change to explain topology. Useful when topology is based on character states, but may not be approriate for sequence based phylogenies. \\
  \\ [-1.5ex]   
  \textbf{Maximum Likelihook} & 
  Add Text Here  \\
  \\ [-1.5ex]  
  \textbf{Bayesian} & 
  Add Text Here  \\
  \\ [-1.5ex]  
  \hline
\end{tabular}
\end{center}

### A. Neighbor Joining Trees
Neighbor joining trees are created from a distance matrix that is generated by comparing pairs of alinged sequences.
Trees are then constructed in an agglomerative fashion, similar to the way that we conducted cluster analysis in the $\beta$-diversity exercise.
The tree-building starts with an unresolved tree in "star network".
The neighbor joining algorithm then identifies the two most similar taxa and creates a node.
This process is iteratively applied to the remaining sequences until the tree is resolved.  

The first step in making a neighbor joining tree is to create a distance matrix. 
We will do this with the `dist.dna` function in the `ape` package. 
In the R chunk below, the "model = raw"" argument means that `dist.dna` will estimate distances based on the proportion of sites that differ between pairs of sequences.
The "pairwise.deletion = FALSE" argument means that an entire site (i.e., column) is deleted if there is a missing observation (i.e., a gap) for one or more of the sequences (i.e., rows). 

```{r}
# Create distance matrix with "raw" model {ape}
seq.dist.raw <- dist.dna(p.DNAbin, model = "raw", pairwise.deletion = FALSE)
```

After calculating the distances between our sequences, we are now ready to make a neighor joining tree using the `nj` function in `ape`. 
In the process, we will identify an outgroup sequence (*Methanosarcina*) that we will use to root the tree. 

```{r}
# Neighbor Joining Algorithim to Construct Tree, a 'phylo' Object {ape}
nj.tree <- nj(seq.dist.raw)

# Identify Outgroup Sequence
outgroup <- match("Methanosarcina", nj.tree$tip.label)

# Root the Tree {ape}
nj.rooted <- root(nj.tree, outgroup, resolve.root = TRUE)

# Plot the Rooted Tree {ape}
plot.phylo(nj.rooted, main = "Neigbor Joining Tree", "phylogram", use.edge.length = FALSE,
           direction = "right", cex = 0.7, label.offset = 1)
add.scale.bar(cex = 0.7)
```

### B) SUBSTITUTION MODELS OF DNA EVOLUTION 
Phylogenetic analyses require assumptions about the rates of DNA evolution. 
Various models have been developed to estimate the rates that nucleotides change over time. 
These models are important for inferring relatedness of genes and/or taxa.

###Perhaps convert text below into table. 
The **Jukes-Cantor model (JC69)** is the simplest **substitution model**.
It assumes that all nucleotides occur at equal frequencies and that these nucleotides can mutate from one to another with equal probability. 
The **Felenstein model (F81)** builds from JC69 by allowing base frequences to vary. 
The **Kimura model (K80)** also assumes equal frequencies of nucleotides, but recognizes that transition mutations (e.g., purine to a purine [A -> G] or pyrmidine to pyrmidine [C -> T]) occur with higher probability than tranversion mutations (e.g., purine to pyrmidine or vice-versa).
**Hasegawa, Kishino, and Yano (HKY85)** merges elements of F81 and K80 by accounting for uneuqal base frequencies and differences in the rates of transitions and tranversions. 
Many other models of DNA evolution exist.
For example, the **Tamura (T92)** accounts for G+C bias, while **General Time Reversible (GTR)** models capture the probabilities associated with nucleotide reversions (e.g., T -> C -> T).


\begin{center}
\hyphenpenalty 10000
\exhyphenpenalty 10000
\begin{tabular}{ m{5cm} m{10cm} }
  \textbf{Evolutionary Model} & \textbf{Properties} \\
  \hline \hline \\ [-1.5ex]
  \textbf{Jukes-Cantor model (JC69)} & 
  the simplest **substitution model**. It assumes that all nucleotides occur at equal frequencies and that these nucleotides can mutate from one to another with equal probability.  \\
  \\ [-1.5ex]
  \textbf{Felenstein model (F81)} & 
  builds from JC69 by allowing base frequences to vary.  \\
  \\ [-1.5ex]  
  \textbf{Kimura model (K80)} & 
  assumes equal frequencies of nucleotides, but recognizes that transition mutations (e.g., purine to a purine [A -> G] or pyrmidine to pyrmidine [C -> T]) occur with higher probability than tranversion mutations (e.g., purine to pyrmidine or vice-versa).  \\
  \\ [-1.5ex]   
  \textbf{Hasegawa, Kishino, and Yano (HKY85)} & 
  merges elements of F81 and K80 by accounting for uneuqal base frequencies and differences in the rates of transitions and tranversions. \\
  \\ [-1.5ex]   
  \hline
\end{tabular}
\end{center}
  
Fortunately, the `dist.dna` function that we used above for generating the neighbor joining tree can easily create distance matrices for a large number of DNA substitution models. 
Let's create a distance matrix among our bacterial isolates based on the Jukes-Cantor (JC69) subsitution model using the `dist.dna` function in `ape`. 

```{r}
# Create distance matrix with "JC69" model {ape}
seq.dist.JC69 <- dist.dna(p.DNAbin, model = "JC69", pairwise.deletion = FALSE)
```

Now, let's compare the "raw" and "JC69" distance matrices. 
First, we will make a **saturation plot**, which allow us to visuazlie the effect of our DNA subsititution model compared to a distance matrix that does not include this feature. 

```{r}
# Plot Distances from Different DNA Substitution Models
plot(seq.dist.raw, seq.dist.JC69, 
     pch = 20, col = "red", las = 1, asp = 1, xlim=c(0,0.7), ylim = c(0,0.7),
     xlab = "Raw Distance", ylab = "JK69 Distance") 
abline(b =1, a = 0, lty = 2)
text(0.65, 0.6, "1:1")
```

Second, we will assess how the "raw" and "K80" distance matrices affect tree topology by making a **cophylogenetic plot**. 
This will give us a visual sense of whether or not the two trees are in agreement with one another. 

```{r}
# Make Neighbor Joining Trees Using Different DNA Substitution Models {ape}
raw.tree <- nj(seq.dist.raw)
JC69.tree <- nj(seq.dist.JC69)

# Define Outgroups
raw.outgroup <- match("Methanosarcina", raw.tree$tip.label) 
JC69.outgroup <- match("Methanosarcina", JC69.tree$tip.label) 

# Root the Trees {ape}
raw.rooted <- root(raw.tree, raw.outgroup, resolve.root=TRUE) 
JC69.rooted <- root(JC69.tree, JC69.outgroup, resolve.root=TRUE)

# Make Cophylogenetic Plot {ape}
layout(matrix(c(1,2), 1, 2), width = c(1.2, 1))
par(mar = c(4, 1, 2, 0.1))
plot.phylo(JC69.rooted, type = "phylogram", direction = "right", 
           use.edge.length = FALSE, adj = 0.5, cex = 0.7, label.offset = 1, main = "JC69 tree")

axisPhylo()
par(mar = c(4, 0, 2, 1))
plot.phylo(raw.rooted, type = "phylogram", direction = "left", 
           use.edge.length = FALSE, show.tip.label = FALSE, main = "raw tree")
axisPhylo()
layout(1)
```

***Question 3***:  Using the saturation plot and cophylogenetic plot, describe the effect that that the JC69 substitution model has on our phylogenetic reconstruction. If the plots seem to be inconsistent with one another, explain what it giving rise to the differences.   

> ***Answer 3***:   

### C.  MAXIMUM LIKELIHOOD
Maximum likelihood is an approach for estimating parameters for a statistical model.
It is also one of the most commonly used methods for phylogenetic reconstruction.
In this exercise, we will use maximum likelihood to identify DNA substitution models that best fit our sequence data. 
We will then use one of these models to test phylogenetic hypotheses about the distribution of traits among the bacteria in our study system. 

* following a lot of what's on following links: 
* http://cran.r-project.org/web/packages/phangorn/vignettes/Trees.pdf##
* http://cran.r-project.org/web/packages/phangorn/vignettes/Trees.pdf
*  http://www.slideshare.net/KlausVigo/phylogenetics-cibio

* need to think about the "exclude = pairwise" command in the dist.ml function below. 
*  also need to get a better handle on what `optim.pml` is doing
*  and also GTR. 
*  should have a few sentence desription here saying that we're going to take a peak at a few different substitution model, before more on to investigating all. 

### i.  Comparing Two Models 
In the previous section, we used saturation plots and cophylogenetic trees to visualize the effect of a DNA substitution model on our data. 
Here, we will test whether or not the K80 phylogeny is better than the raw phylogeny using maximum likelihood. 
```{r}
# Read in FASTA-Aligned File as a Matrix
read.aln.mat <- as.matrix(read.aln) 

# Converts FASTA-Aligned File into a phyDat object {phangorn}
read.aln.phydat <- phyDat(read.aln.mat, type = "DNA") 

# Make Neighbor Joining Tree via Pairwise Distance for phyDat Object {phangorn}
tree.raw.mle <- nj(dist.logDet(read.aln.phydat)) 

# Identify Outgroup
outgroup <- match("Methanosarcina", tree.raw.mle$tip.label)

# Root Tree {ape}
rooted.raw.mle <- root(tree.raw.mle, outgroup, resolve.root=TRUE)

# Maximum Likelihood function {phangorn} returns tree, data, output
tree.JC.mle <- pml(rooted.raw.mle, read.aln.phydat) 

# Optimizes model parmameters {phangorn}
tree.JC.opt <- optim.pml(tree.JC.mle, TRUE) 
```

Now let's explore some of the output from the maximum liklihood analysis to see which of the few models best fits our data. 

```{r}
# anova(tree.raw.mle, tree.JC.mle, tree.JC.opt)
# AIC(fit.JC, fit.JC.opt, fit.GTR, fit.GTR.opt)
# SH.test(fit.JC, fit.JC.opt, fit.GTR, fit.GTR.opt)
```

### ii.  Identifying the "Best" Model
Now let's evaluate all of the DNA substitution models available to us and see which one best fits our data. 
This analysis is somewhat computationally intensive. 
It may take a couple of minutes to run. 
Be patient.
When this is done, we can see which DNA substituion model maximum likelihood determined to be the best.

```{r}
model.tests <- modelTest(read.aln.phydat) # {phangorn}
sort.tests <- model.tests[order(model.tests$AIC),] # order by AIC
best.model <- sort.tests[1,1]
env <- attr(model.tests, "env")
fitStart <- eval(get(model.tests$Model[which.min(model.tests$AIC)], env), env)

# outgroup <- match("Methanosarcina", fitStart$tip.label)
# fit.best.rooted <- root(fitStart, outgroup, resolve.root=TRUE)
# fitStart.rooted <- root(fitStart, outgroup, resolve.root=TRUE) # reroots the tree

fit.best <- optim.pml(fitStart, optNni=TRUE, optGamma=TRUE, optInv=TRUE, model="GTR")
# fit.best.rooted <- root(fit.best, outgroup, resolve.root=TRUE)
# bs <- bootstrap.pml(fit.best, bs=2, optNni=TRUE, multicore=TRUE) ### took long time (10 mins)
# par(mar=c(.1,.1,.1,.1))
# plotBS(fit.best$tree, bs) #{phangorn}
```

### C. Obtaining Bootstrap Support for the Tree
Describe what bootstrapping is, how it's done, and how to interpret. 
Somewhat slow process (2 mins)
Read up on arguments

```{r}
# bs <- bootstrap.pml(fit.JC, bs = 100, optNni = TRUE,
#             control = pml.control(trace = 0)) #{phangorn}  
```

### Make tree with bootstrap support values 
There is apparently a problem here. plotBS hangs forever

```{r}
# par(mar=c(.1,.1,.1,.1))
# plotBS(fit.JC$tree, bs) #{phangorn}
```

http://grokbase.com/t/r/r-help/12876aby2z/r-how-to-write-out-a-tree-file-with-bootstrap-from-phangorn-package
"tree <- plotBS(fittree, bs)
You can export than this tree using write.tree or write.nexus, e.g. write.tree(tree)"

## 8) INTEGRATING TRAITS AND PHYLOGENY -- OVERVIEW, CATEGORICAL, CONTINUOUS
Now that we have custructued of phylogeny, we can move on to incorporating traits into our analysis.


### A. Loading Trait Database
```{r}
# Import Growth Rate Data
p.growth <- read.table("./data/p.isolates.raw.growth.txt", sep = "\t", header = TRUE, row.names = 1)

# Standadize Growth Rates
p.growth.std <- p.growth / (apply(p.growth, 1, sum))

# Calculate Max Growth Rate
umax <- (apply(p.growth, 1, max)) 
```

### B. Calculating Niche Breadth
We calcualte the niche breadth (*nb*) index from Levins (1968) which is defined as $\frac{1}{\left( n \cdot \Sigma p_{xi}^{2} \right)}$, where ni is each resource 

```{r}
levins <- function(p_xi = ""){
  x = 0
  for (i in p_xi){
    x = x = i^2
    }
  nb = 1 / (length(p_xi) * x)
  return(nb)
  }
```

Now that we have a function to calculate niche breadth, we can calculate the niche breadth of our isolates on phosphorus resources.


```{r}
nb <- levins(p.growth.std)
```  

### C. Visualzing Traits on Trees
MULTIPLE P TRAITS ON A TREE
SINGLE TRAIT (NICHE BREADTH) ON TREE

```{r}

## Why using the bionj function?  how about `nj`
# nj.tree <- nj(seq.dist) # ape command for Neighbor Joining
# outgroup <- match("Methanosarcina", nj.tree$tip.label)
# nj.rooted <- root(nj.tree, outgroup, resolve.root=TRUE)
# nj.rooted <- drop.tip(nj.rooted, "Methanosarcina")

# Using the Best Fit Model
bs <- bootstrap.pml(fit.best)
bs.1 <- bs[[1]]


outgroup <- match("Methanosarcina", bs.1$tip.label)
fit.best.rooted <- root(bs.1, outgroup, resolve.root=TRUE)
fit.best.noroot <- drop.tip(fit.best.rooted, "Methanosarcina")


traits <- p.growth.std[,c("ATP", "SRP", "PolyP", "Phyt", "AEP")]
trait.nb <- as.data.frame(nb)
rownames(trait.nb) <- rownames(traits)

# Drop Missing Branches and data
fit.best.noroot <- drop.tip(fit.best.noroot, c("LL11A", "LL15", "WG37", "WG30A", "WG32", "LL12A"))

require(phylobase)
require(adephylo)
x <- phylo4d(fit.best.noroot, traits)
table.phylo4d(x, treetype="clad", symbol="colors", show.node=TRUE, 
              cex.label = 0.75, scale = TRUE, use.edge.length=FALSE, 
              edge.color="black", edge.width=2, box=FALSE, 
              col=topo.colors(10), pch = 15, cex.symbol = 1, 
              ratio.tree=0.75, cex.legend=1)                                                                                                                                                                

# Niche Breadth
x.nb <- phylo4d(fit.best.noroot, trait.nb)
table.phylo4d(x.nb, treetype="clad", symbol="colors", show.node=TRUE, 
              cex.label = 0.75, scale = TRUE, use.edge.length=FALSE, 
              edge.color="black", edge.width=2, box=FALSE, 
              col=topo.colors(10), pch = 15, cex.symbol = 1, 
              ratio.tree=0.75, cex.legend=1) 

# Still need to clean it up. bUt looking much better

# library(latticeExtra)
# dd.col <- as.dendrogram(hc)
# col.ord <- order.dendrogram(dd.col)
# mat <- matrix(rnorm(4*4),nrow=4)
# colnames(mat) <- tree$tip.label
# rownames(mat) <- tree$tip.label
# levelplot(mat[tree$tip,tree$tip],type=c('g','p'),
#           aspect = "fill",
#           colorkey = list(space = "left"),
#           legend =
#             list(right =
#                    list(fun = dendrogramGrob,
#                         args =
#                           list(x = dd.col, 
#                                side = "right",
#                                size = 10))),
#           panel=function(...){
#             panel.fill('black',alpha=0.2)
#             panel.levelplot.points(...,cex=12,pch=23)
#           }
# )
```

## 9) HYPOTHESIS TESTING

### A) PAGEL'S LAMBDA - Jay
```{r}
fitContinuous(fit.best.noroot, trait.nb)
```



### B) BOOMBERG'S K - Jay


### C) DISPERSION "D" --> CAPER PACKAGE? CONTINUOUS VS. CATEGORICAL
    if this doesn't work, can we hijack phylo.struct from picante
    -- Mario
### D) CONSTRAINT: 
http://ps.uci.edu/scholar/amartiny/files/consentrait.txt
  -- Mario
### E) Phylogenetic independent contrasts -- Jay
  

## 10) HOMEWORK
Test growth rate-niche breadth trade-off using pylognetic independent contrasts. Still signficant?



Phylogenetic signal occurs when the traits of closely related species are more similar than what is expected by chance. 
In many cases, testing for phylogenetic signal is a prerequsite for addressing questions related to trait evolution and phylogenetic biodiversity. 
If there is significant phylogenetic signal, then evolutionary history is important for understanding the relationships between for example a trait and the peformance of individuals being considered. 
If there is *not* any phylogenetic signal, then the relationship between traits and peformance can explored using normal statistical methods. 

### Pagelâ€™s Lambda
http://bodegaphylo.wikispot.org/iv._testing_phylogenetic_signal_in_r
http://blog.phytools.org/2013/10/simplified-lambdatree.html
```{r}
# traits.1 <- traits[,1,drop=FALSE]
# fitContinuous(nj.rooted, traits.1)

```

## 9) PHYLOGENETIC INDEPENDENT CONTRASTS
Phylogenetic independent contrasts are outlined in Paradis (2012) on pp 204-209
Can we take a trait-peformance relationship that's siginficant and correct for phylogeny? 
Is the relationship still significant? 
Do the parameters or explained varition change?
```{r}
# pic.nb <- pic(trait.nb, fit.best.noroot)
# pic.umax <- pic(umax, fit.best.noroot)
# 
# fit.best.noroot$tip.label
```


