---
title: "Phylogenetic Diversity - Traits"
author: "Z620: Quantitative Biodiversity, Indiana University"
date: "February 20, 2015"
header-includes:
   - \usepackage{array}
output: pdf_document
geometry: margin=2.54cm
---

## OVERVIEW
Up to this point, we have been focusing on patterns taxonomic diversity in Quantitative Biodiversity. 
Although taxonomic diversity is an important dimension of biodiversity, it does not consider the evolutionary history and relatedness of species. 
The goal of this exercise is to introduce basic concepts of phylogenetic diversity. 

After completing this exercise you will know how to:

1.  create phylogenetic trees
2.  map traits onto phylogenetic trees
3.  test for phylogenetic signal of traits on a phylogenetic tree
4.  test for relationships among traits using phylogenetic independent contrasts 

## 1) SETUP
### A. Retrieve and Set Your Working Directory

```{r, results = 'hide'}
rm(list = ls())
getwd()
setwd("~/GitHub/QuantitativeBiodiversity/Assignments/PhyloTraits")
```

### B. Load Packages 
There are many contributed R packages that have been developed for conducting phylogenetic analyses (http://goo.gl/DtU16j). 
We will rely heavily on the `ape` (http://goo.gl/nExs0) and `phangorn` (http://goo.gl/iWU6Vt) packages, which peform a wide range of phylogenetic analyses. 
We will supplement this with other packages for some specialized tasks such as sequence alignment (`muscle`), reading of alinged sequences (`seqinr`), and testing for phylogenetic signal (`geiger`).
After the initial installation of these packages using the `install.packages()` fucntion, let's load the packages and their dependencies with the `require()` function:

```{r, results = 'hide', message = FALSE, warning = FALSE} 
require("muscle")
require("ape")
require("seqinr")
require("phangorn")
require("phylobase")
require("adephylo")
require("geiger") 
```

## 2) DESCRIPTION OF DATA
The maintenance of biodiversity is thought to be influenced by **trade-offs** among species in certain functional traits. 
One such trade-off involves the ability of a highly specialized species to peform exceptionally well on a particular resource. 
This strategy may be advantageous if the specialist lives in an environment where its preferred resource is found in a relatively high and constant supply. 
However, the specialist strategy may not be optimal if the species finds itself itself in an environment where resource conditions are more variable.
Under these conditions, it may be beneficial if a species is capable of using a wider range of resources, even though there may be a fitness cost associated with being a generalist. 
We set out to test this hypothesized trade-off in bacteria by assessing generalist and specialist strategies on different forms of phopshorus, a resource which commonly limits growth and reproduction of organisms inhabiting freshwater ecosystems.
We collected 39 strains of bacteria from two lakes. 
After identifying these bacteria based on their 16S rRNA gene sequences, we measured their growth on 18 different forms of phosphorus. 
We will take a phylogenetic approach to mapping phosphorus resource use onto a phylogenetic tree while testing for specialist-generalist trade-offs. 

## 3) SEQUENCE ALIGNMENT
After generating or retrieving sequence data, the first step of a phylogenetic analysis is to peform an alignment. 
This can be done with nucleotides (DNA or RNA) or proteins (amino acids). 
There are various methods for aligning sequences, but the basic premise is to use an algorithm to arrange sequences relative to one another based on conserved regions so that downstream phylogenetic analyses can be performed.
In R, there are at least two packages that can be used for aligning sequences: `ape` and `muscle`. 
In `ape`, alignments can be done with the functions `clustal`, `muscle`, and `tcoffee`, all which have a range of options. 
However, these functions require that other programs be installed on your computer.
Therefore, for this exericse, we are going to use the `muscle` package in R becasue it will align sequences in a "stand alone" fashion without the need for installing additional software. 

### A.  Examining a FASTA-File 
We will read in a **FASTA-formatted** file into the `muslce` alignment function.
This is a very common text-based format that is used across platforms and disciplines. 
Let's take a second to look at the *p.isolates.fasta* file. 
Open the terminal and change directory to the data folder. 
You can then type "nano p.isolates.fasta" at the command line. 
This will open the FASTA file in nano, which is a text editing program. 
You can use Ctrl+V and Ctrl+Y to scroll towards the top and bottom of the file, respectively.
When you are done, you can close the file by hitting Ctrl+X.
Do not save any inadvertent changes to *p.isolates.fasta*.

### B.  Performing an Alignment 
Now, let's align our sequences. 
Using the `muscle` fuction, we will read in our FASTA file and create a **FASTA-aligned* output file (with an .afa extension).  

```{r}
# Identify input and output files to perform alignment in the `muscle` package
muscle::muscle(seqs="./data/p.isolates.fasta", out = "./data/p.isolates.afa") 
```

***Question 1***: Using nano or your favorite text editor, compare the `p.isolates.fasta` file and the `p.isolates.afa` file. 
Describe the differences that you observe between the files. 

> ***Answer 1***: 

### C.  Visualizing the Alignment
It is good practice to view your sequences after alignment. 
This may not be practical for large databases, but in our example, the alignment only contains sequences for 39 bacterial species plus a distantly related reference sequence that we will use as an **outgroup**. 
To visualize our alingment, first, we will read in the FASTA-aligned file using the `read.alignment` function in the `seqinr` package. 
Then, using the `ape` package, we will convert the alignnment file it into a **DNAbin object**, which is a bit-level coding scheme that allows R to store and manipulate sequence data.
Last, we will visualize the alignment by color-coding nucelobases different colors, where red = adenine, yellow = guanine, green = cytosine, blue = thymidine, grey = ambiguous call, and black = alignment gaps. 

```{r}
# Read Alignment File {seqinr}
read.aln <- read.alignment(file = "./data/p.isolates.afa", format = "fasta")  

# Convert Alignment File to DNAbin Object {ape}
p.DNAbin <- as.DNAbin(read.aln) 

# Identify Base Pair Region of 16S rRNA Gene to Visuzlize
window <- p.DNAbin[,100:500] 

# Command to Visusalize Sequence Alignment {ape}
image.DNAbin(window, cex.lab = 0.6) 

# Optional Code Adds Grid to Help Visualize Rows of Sequences 
grid(ncol(window), nrow(window), col = "lightgrey") 
```

***Question 2***:  Make some observations about the `muscle` alingmnet of the 16S rRNA gene sequences for our bacterial isolates and the outgroup, *Methanosarcina*, which belongs to the archaeal domain of life. 
Move along the alignment by changing the values in the `window` object.  
a.  Approximatley how long were our reads?  
b.  What regions do you think would be most appropriate for phylogenetic inference?  

> ***Answer 2a***:   
> ***Answer 2b***:  

## 4) MAKING A PHYLOGENETIC TREE
Once you have aligned your sequences, the next step is to construct a phylogenetic tree.
Not only is this effective for visualizing the evolutionary relationship among taxa, but as you will see later, the information that goes into a phylogenetic tree is needed for downstream hypothesis testing. 

We could spend an entire semester discussing the theory and practice behind making phylogenetic trees.
The objective of this handout is to quickly get you up to speed on some of the basic principles of phylognetic reconstruction. 
In adition, we want to convey some of the opportunities and limitations of conducting phylogenetic analyses in R. 
The following table provides an overview of the methods that are available for evolutionary reconstruction in R. 

\begin{center}
\hyphenpenalty 10000
\exhyphenpenalty 10000
\begin{tabular}{ m{5cm} m{10cm} }
  \textbf{Tree Method} & \textbf{Properties} \\
  \hline \hline \\ [-1.5ex]
  \textbf{UPGMA} & 
  Hierarchical clustering technique. Progressively groups the most similiar pairs of sequences. Sometimes used as a guide tree for more sophisticated phylogenetic analyses, but prone to distorting tree topology and assumes constant rates of evolution. UPGMA can be implemented in R with the upgma function in the phangorn package.\\
  \\ [-1.5ex]
  \textbf{Neighbor Joining} & 
  Starts with an unresolved tree in a "star network". Algorithm then identifies the two most similar taxa and creates a node. This process is iteratively applied to the remaining sequences. Can serve as a starting point for assessing evolutionary models with more sophisticated methods like maximum likelihood. Neighbor joining can be implemented in the R packages `ape` and phangorn using the function nj \\
  \\ [-1.5ex]  
  \textbf{Maximum Parsimony} & 
  A character-based, non-parametric approach that involves identifying one of many possible trees based upon the minimum number of evolutionary "steps"" (or mutations). No underlying evolutionary model. Maximum parsimony can be executed using the `parsimony` function in the package panghorn. The function optim.parsimony performs tree rearrangements using nearest-neighbor interchanges (NNI) to optimize parsimony scores.\\
  \\ [-1.5ex]   
  \textbf{Maximum Likelihood} & 
  Parameteric approach to identifying the relative probability associated with data fitting a given model of character evolution. One of the most commonly used approaches for building trees from sequence data. Statistically sound and flexible for accomodating realistic models of sequence evolution. Maximum likelihood can be implemented in R using the `pml` function in the pranghorn package. Parameters can be optimized using the optim.pml and evolutionary models can be compared with the model.tests function\\
  \\ [-1.5ex]  
  \textbf{Bayesian} & 
  Uses likelihood function and prior probabilities to create a posterior probability of a tree fitting a given model of evolution. Computationally intensive; often uses Markov chains Monte Carlo (MCMC). Unlike other methods, parameters are not constant; rather, they come from probabilistic distributions. Along with maximum likelihood, considered one of the more preferable ways of generating trees. R has some packages (e.g., ips) the wrap popular Bayesian software such as MrBayes\\
  \\ [-1.5ex]  
  \hline
\end{tabular}
\end{center}

### A. Neighbor Joining Trees
For the purposes of this course, we are going to focus primarily on building trees using neighbor joining methods.
As you will see in the next section, howevever, neighbor joining is commonly used for making "guide tree" that can incorporate more sophisticated models of evolution. 
The first step in making a neighbor joining tree is to create a distance matrix. 
We will do this with the `dist.dna` function in the `ape` package. 
In the R chunk below, the "model = raw"" argument means that `dist.dna` will estimate distances based on the proportion of sites that differ between pairs of sequences.
The "pairwise.deletion = FALSE" argument means that an entire site (i.e., column) is deleted if there is a missing observation (i.e., a gap) for one or more of the sequences (i.e., rows). 

```{r}
# Create distance matrix with "raw" model {ape}
seq.dist.raw <- dist.dna(p.DNAbin, model = "raw", pairwise.deletion = FALSE)
```

After calculating the distances between our sequences, we are now ready to make a neighor joining tree using the `nj` function in `ape`. 
In the process, we will identify an outgroup sequence (*Methanosarcina*) that we will use to root the tree. 

```{r}
# Neighbor Joining Algorithim to Construct Tree, a 'phylo' Object {ape}
nj.tree <- nj(seq.dist.raw)

# Identify Outgroup Sequence
outgroup <- match("Methanosarcina", nj.tree$tip.label)

# Root the Tree {ape}
nj.rooted <- root(nj.tree, outgroup, resolve.root = TRUE)

# Plot the Rooted Tree{ape}
plot.phylo(nj.rooted, main = "Neigbor Joining Tree", "phylogram", use.edge.length = FALSE,
           direction = "right", cex = 0.7, label.offset = 1)
add.scale.bar(cex = 0.7)
```

### B) SUBSTITUTION MODELS OF DNA EVOLUTION 
Phylogenetic analyses require assumptions about the rates of DNA evolution. 
Various models have been developed to estimate the rates that nucleotides change over time. 
These models are important for inferring relatedness of genes and/or taxa.
In the following table, we describe a few of the more commonly used models.  

\begin{center}
\hyphenpenalty 10000
\exhyphenpenalty 10000
\begin{tabular}{ m{5cm} m{10cm} }
  \textbf{Evolutionary Model} & \textbf{Properties} \\
  \hline \hline \\ [-1.5ex]
  \textbf{Jukes-Cantor model (JC69)} & 
  Simplest model. Assumes all nucleotides occur at equal frequencies and that these nucleotides can mutate from one to another with equal probability.\\
  \\ [-1.5ex]
  \textbf{Felenstein model (F81)} & 
  Builds from JC69 by allowing nucleotide frequencies to vary.\\
  \\ [-1.5ex]  
  \textbf{Kimura model (K80)} & 
  Assumes equal frequencies of nucleotides, but recognizes that transition mutations (e.g., purine to a purine [A -> G] or pyrmidine to pyrmidine [C -> T]) occur with higher probability than tranversion mutations (e.g., purine to pyrmidine or vice-versa).\\
  \\ [-1.5ex]   
   \textbf{Felenstein model (F84)} & 
  Assumes different rates of base transitions and trasnversions while allowing for differences in base frequencies.\\
  \\ [-1.5ex] 
  \textbf{Tamura model (T92)} & 
  Similar to K80 but accounts for G + C content.\\
  \\ [-1.5ex]  
  \textbf{General Time Reversible model (GTR)} & 
  Capture the probabilities associated with nucleotide reversions (e.g., T -> C -> T). All substitution rates are different. Does not assume equal base frequencies.\\
  \\ [-1.5ex] 
  
  \hline
\end{tabular}
\end{center}
  
Fortunately, the `dist.dna` function that we used above for generating the neighbor joining tree can easily create distance matrices for a large number of DNA substitution models. 
Let's create a distance matrix among our bacterial isolates based on the Fellenstein model (F84) subsitution model using the `dist.dna` function in `ape`. 

```{r}
# Create distance matrix with "F84" model {ape}
seq.dist.F84 <- dist.dna(p.DNAbin, model = "F84", pairwise.deletion = FALSE)
```

Now, let's compare the "raw" and "F84" distance matrices. 
First, we will make a **saturation plot**, which allow us to visuazlie the effect of our DNA subsititution model compared to a distance matrix that does not include this feature. 

```{r}
# Plot Distances from Different DNA Substitution Models
plot(seq.dist.raw, seq.dist.F84, 
     pch = 20, col = "red", las = 1, asp = 1, xlim=c(0,0.7), ylim = c(0,0.7),
     xlab = "Raw Distance", ylab = "F84 Distance") 
abline(b =1, a = 0, lty = 2)
text(0.65, 0.6, "1:1")
```

Second, we will assess how the "raw" and "F84" distance matrices affect tree topology by making a **cophylogenetic plot**. 
This will give us a visual sense of whether or not the two trees are in agreement with one another. 

```{r}
# Make Neighbor Joining Trees Using Different DNA Substitution Models {ape}
raw.tree <- nj(seq.dist.raw)
F84.tree <- nj(seq.dist.F84)

# Define Outgroups
raw.outgroup <- match("Methanosarcina", raw.tree$tip.label) 
F84.outgroup <- match("Methanosarcina", F84.tree$tip.label) 

# Root the Trees {ape}
raw.rooted <- root(raw.tree, raw.outgroup, resolve.root=TRUE) 
F84.rooted <- root(F84.tree, F84.outgroup, resolve.root=TRUE)

# Make Cophylogenetic Plot {ape}
layout(matrix(c(1,2), 1, 2), width = c(1.2, 1))
layout.show(2)
par(mar = c(4, 1, 2, 0))
plot.phylo(F84.rooted, type = "phylogram", direction = "right", show.tip.label=TRUE,
           use.edge.length = FALSE, adj = 0.75, cex = 0.7, label.offset = 2, main = "F84 tree",  )

axisPhylo()
par(mar = c(4, 0, 2, 1))
plot.phylo(raw.rooted, type = "phylogram", direction = "left", show.tip.label=TRUE,
           use.edge.length = FALSE, , main = "raw tree", label.offset = 0)
axisPhylo()
layout(1)

cophyloplot(F84.rooted, raw.rooted, assoc=TRUE)

```

***Question 3***:  Using the saturation plot and cophylogenetic plot, describe the effect that that the JC69 substitution model has on our phylogenetic reconstruction. If the plots seem to be inconsistent with one another, explain what it giving rise to the differences.   

> ***Answer 3***:   

## 5) INTEGRATING TRAITS AND PHYLOGENY
In the context of biodiveristy, many researchers are interested in being able to map traits onto phylogenetic trees.
This can be important for addressing questions related to diversification, trait evolution, and ecological interactions.
In this section, we import functional trait data that was collected on the 39 freshwater isoaltes for which we have 16S rRNA sequences. 
Specifically, we measured the growth rates of each strain on 18 different forms of phosphorus (e.g., phosphate, DNA, ATP, etc.). 
With this information we will visualize the functional traits (i.e., growth on different phosphorus) in a phylogenetic conetxt by mapping them onto the neighbor joining tree (with the F84 DNA substituion model) that we created. 

### A. Loading Trait Database
```{r}
# Import Growth Rate Data
p.growth <- read.table("./data/p.isolates.raw.growth.txt", sep = "\t", header = TRUE, row.names = 1)

# Standadize Growth Rates Across Strains
p.growth.std <- p.growth / (apply(p.growth, 1, sum))
```

### B. Calculating Variables to Test for a Generalist-Specialist Trade-off
From the section above where we described the data, you will recall that we were interested in testing for a **generalist-specialist** trade-off.
To do this, we need to create two new variables. 
First, we need calculate the maximum growth rate of each isoalte across all phosphorus types. 
This will help us test the expectation that generalists will have lower maximum growth rates because there is a cost associated with being able to use a lot of different chemical forms of phosphorus. 
In contrast, we expect that specialists will have a high maximum growth rate on their preferred phosphorus resource. 

```{r}
# Calculate Max Growth Rate
umax <- (apply(p.growth, 1, max)) 
```

Second, we need to come up with a way of quantifying whether a strain is a generalist or a specialist. 
We will used the niche breadth (*nb*) index from Levins (1968) which is defined as $\frac{1}{\left( n \cdot \Sigma p_{xi}^{2} \right)}$, where $n_{i}$ is each resource 

```{r}
levins <- function(p_xi = ""){
  x = 0
  for (i in p_xi){
    x = x + i^2
    }
  nb = 1 / (length(p_xi) * x)
  return(nb)
  }
```

Now let's apply our niche breadth function to the standardized growth rates of our isolates on the different phosphorus resources.

```{r}
nb <- levins(p.growth.std)
```  

### C. Visualzing Traits on Trees


```{r}
# Generate Neighbor Joining Tree Using F84 DNA Substitution Model {ape}
nj.tree <- bionj(seq.dist.F84) # no negative lengths with bionj - improved alg

# Define the Outgroup
outgroup <- match("Methanosarcina", nj.tree$tip.label)

# Create a Rooted Tree {ape}
nj.rooted <- root(nj.tree, outgroup, resolve.root=TRUE)
 
# Keep Rooted but Drop Outgroup Branch
nj.rooted <- drop.tip(nj.rooted, "Methanosarcina")

# Rescale for Negative Branch Length
# nj.rooted$edge.length <- nj.rooted$edge.length + 0.1

# Subset Growth for a Few Phosphorus Sources
traits <- p.growth.std[,c("ATP", "SRP", "PolyP", "Phyt", "AEP")]
trait.nb <- as.data.frame(nb)
rownames(trait.nb) <- rownames(traits)

# Define Color Palett
require(RColorBrewer)
mypalette <- colorRampPalette(brewer.pal(9, "YlOrRd"))


# Need Some More Annotation
x <- phylo4d(nj.rooted, traits)
table.phylo4d(x, treetype="clad", symbol="colors", show.node=TRUE, 
              cex.label = 0.75, scale = TRUE, use.edge.length=FALSE, 
              edge.color="black", edge.width=2, box=FALSE, 
              col=mypalette(100), pch = 15, cex.symbol = 1, 
              ratio.tree=0.75, cex.legend=1, center=FALSE)    

# Need to Fix Color Range
# Niche Breadth
x.nb <- phylo4d(nj.rooted, trait.nb)
table.phylo4d(x.nb, treetype="clad", symbol="colors", show.node=TRUE, 
              cex.label = 0.75, scale = TRUE, use.edge.length=FALSE, 
              edge.color="black", edge.width=2, box=FALSE, 
              col=mypalette(100), pch = 15, cex.symbol = 1, 
              ratio.tree=0.75, cex.legend=1, center=FALSE)

data(mjrochet)
temp <- read.tree(text=mjrochet$tre) # make a tree
mjr <- phylo4d(x=temp,tip.data=mjrochet$tab) # male a phylo4d object
table.phylo4d(x.nb,cex.lab=.5,show.node=FALSE,symb="square")

```
 
# library(latticeExtra)
# dd.col <- as.dendrogram(hc)
# col.ord <- order.dendrogram(dd.col)
# mat <- matrix(rnorm(4*4),nrow=4)
# colnames(mat) <- tree$tip.label
# rownames(mat) <- tree$tip.label
# levelplot(mat[tree$tip,tree$tip],type=c('g','p'),
#           aspect = "fill",
#           colorkey = list(space = "left"),
#           legend =
#             list(right =
#                    list(fun = dendrogramGrob,
#                         args =
#                           list(x = dd.col, 
#                                side = "right",
#                                size = 10))),
#           panel=function(...){
#             panel.fill('black',alpha=0.2)
#             panel.levelplot.points(...,cex=12,pch=23)
#           }
# )
```

## 9) HYPOTHESIS TESTING

### A) PAGEL'S LAMBDA - Jay
```{r}
fitContinuous(fit.best.noroot, trait.nb)
```



### B) BOOMBERG'S K - Jay


### C) DISPERSION "D" --> CAPER PACKAGE? CONTINUOUS VS. CATEGORICAL
    if this doesn't work, can we hijack phylo.struct from picante
    -- Mario
### D) CONSTRAINT: 
http://ps.uci.edu/scholar/amartiny/files/consentrait.txt
  -- Mario
### E) Phylogenetic independent contrasts -- Jay
  

## 10) HOMEWORK
Test growth rate-niche breadth trade-off using pylognetic independent contrasts. Still signficant?



Phylogenetic signal occurs when the traits of closely related species are more similar than what is expected by chance. 
In many cases, testing for phylogenetic signal is a prerequsite for addressing questions related to trait evolution and phylogenetic biodiversity. 
If there is significant phylogenetic signal, then evolutionary history is important for understanding the relationships between for example a trait and the peformance of individuals being considered. 
If there is *not* any phylogenetic signal, then the relationship between traits and peformance can explored using normal statistical methods. 

### Pagelâ€™s Lambda
http://bodegaphylo.wikispot.org/iv._testing_phylogenetic_signal_in_r
http://blog.phytools.org/2013/10/simplified-lambdatree.html
```{r}
# traits.1 <- traits[,1,drop=FALSE]
# fitContinuous(nj.rooted, traits.1)

```

## 9) PHYLOGENETIC INDEPENDENT CONTRASTS
Phylogenetic independent contrasts are outlined in Paradis (2012) on pp 204-209
Can we take a trait-peformance relationship that's siginficant and correct for phylogeny? 
Is the relationship still significant? 
Do the parameters or explained varition change?
```{r}
# pic.nb <- pic(trait.nb, fit.best.noroot)
# pic.umax <- pic(umax, fit.best.noroot)
# 
# fit.best.noroot$tip.label
```


