---
title: "Phylogenetic Diversity - Traits"
author: "Z620: Quantitative Biodiversity, Indiana University"
date: "February 20, 2015"
header-includes:
   - \usepackage{array}
output: pdf_document
geometry: margin=2.54cm
---

## OVERVIEW

## 1) SETUP
### A. Retrieve and Set Your Working Directory

```{r, results = 'hide'}
rm(list = ls())
getwd()
setwd("~/GitHub/QuantitativeBiodiversity/Assignments/PhyloTraits")
```

### B. Load Packages 
```{r, results = 'hide', message = FALSE, warning = FALSE} 
require("muscle")
require("ape")
require("seqinr")
require("phangorn")
```

## 2) DESCRIPTION OF DATA
The meintenance of biodiversity is thought to be influenced by trade-offs among species. 
One such trade-off involves species that are highly specialized at using a single resource. 
However, this may come at the cost of not being able to use other resoruces as well.
We set out to test this hypothesized trade-off in bacteria by assessing generalist and specialist strategies on different forms of phopshorus, which commonly limits growth and reproduction of populations in freshwater ecosystems.
More..

## 3) SEQUENCE ALIGNMENT
It appears that DNA sequences can be alligned in R using at least two packages: `ape` and `muscle`. 
In `ape`, alignments can be done with the functions `clustal`, `muscle`, and `tcoffee`, all which seem to have a range of options. 
However, it seem that these functions call on other installed programs, like ClustalX. 
In contrast, the `muscle` package aligns sequences in a "stand alone" fashion. 

```{r}
seq.aln <- muscle::muscle(seqs="data/p.isolates.fasta", out = "data/p.isolates.afa") 
```

## 4) VISUALIZE THE ALIGNMENT
```{r}
read.aln <- read.alignment(file = "data/p.isolates.afa", format = "fasta") # seqinr command to read muscle align
p.DNAbin <- as.DNAbin(read.aln) # ape command to convert fasta file into DNAbin filetype
x <- p.DNAbin[,100:500] # identify region (in bp of 16S rRNA) to visuzlize
image(x, cex.lab = 0.5) # ape command to visusalize seq align
grid(ncol(x),nrow(x),col="lightgrey") # ape command to add grid
```

## 5) EVOLUTIONARY RECONSTRUCTION
The `ape` package uses the `dist.dna` function to create distances for various DNA substitution models. 
```{r}
seq.dist <- dist.dna(p.DNAbin, model = "K80", pairwise.deletion = TRUE) # ape: Kimuraâ€™s 2-parameters distance
```

## 6) MAKE A NEIGHBOR-JOINING and UPGMA TREE
```{r}
nj.tree <- nj(seq.dist) # ape command for Neighbor Joining
upgma.tree <- upgma(seq.dist)
outgroup <- match("Methanosarcina", nj.tree$tip.label)
nj.rooted <- root(nj.tree, outgroup, node = NULL)
layout(matrix(c(1,2), 2, 1), height=c(1,2))
par(mar = c(.1,.1,.1,.1))
plot.phylo(nj.rooted, main = "NJ", "phylogram", use.edge.length = FALSE, direction = "right", cex = 0.5, label.offset = 1) # ape command
plot(upgma.tree, main = "UPGMA", use.edge.length = FALSE, direction = "right", cex = 0.5, label.offset = 1)
```

## 7) MAKE A PARSIMONY TREE
Using package `phangorn`.
Requires phyDat formatted objects.
There are functions like `optim.parsimony` that peforms tree rearragements via nearest-neighbor exchange and `pratchet` which uses parsimony ratchet to find trees with lower parsimony score.
Type the following at the R terminal for a detailed exmaple: vignette('Trees')

```{r}
read.aln.mat <- as.matrix(read.aln)
read.aln.phydat <- phyDat(p, type = "DNA")
seq.dist.phydat <- dist.dna(as.DNAbin(read.aln.phydat))
nj.tree.phydat <- nj(seq.dist.phydat)
parsimony(nj.tree.phydat, read.aln.phydat)
```

## 6) MAXIMUM LIKELIHOOD ESTIMATION AND BOOTSTRAP SUPPORT
**We may need to use the package `phangorn` (http://cran.r-project.org/web/packages/phangorn/phangorn.pdf)**
** Discussed by Paradis (2012) on pp 138-159
```{r}
ml <- pml(nj.tree.phydat, data = seq.dist.phydat) 

seq.dist.phydat

## 7) MAPPING TRAITS ONTO TREES
** Discussed by Paradis (2012), chapter 4
** Basically, want to get to the point where people can visusalize singe or multiple traits mapped onto brances of trees
** Could do this with Kali's P isoltes or with Lennon et al. (2010) with soil isolates and traits. 

## 8) PHYLOGENETIC SIGNAL
**Phylogenetic signal is outlined in Paradis (2012) on pp 236-239
**Can we take a big-picture look at the phylogeny and traits combined and deteremine whether or not their is "signal" beyond null Brownian motion?
Hopefully, we pick an example where the answer is "yes"
** visualig co-mapping, mantel tests

## 9) PHYLOGENETIC INDEPENDENT CONTRASTS
**Phylogenetic independent contrasts are outlined in Paradis (2012) on pp 204-209
**Can we take a trait-peformance relationship that's siginficant and correct for phylogeny? 
Is the relationship still significant? 
Do the parameters or explained varition change?

