---
title: "Geographical Ecology"
author: "Z620: Quantitative Biodiversity, Indiana University"
date: "February 13, 2015"
header-includes:
   - \usepackage{array}
output: pdf_document
geometry: margin=2.54cm
---

## OVERVIEW
In this exercise, we expand upon concepts $\beta$-diversity through a more explicit treatment of geographic ecology. 

### Can we mention something up here about species area relationships; something more concrete?
###adressing marjor eco concepts?##

After completing this exercise you will know how to:

1. Quantify the aggregation of species
2. Visualize geographical patterns of species 
3. Test hypotheses regarding the spatial correlation of species


## 1) SETUP
### Retrieve and Set Your Working Directory

```{r, results = 'hide'}
rm(list=ls())
getwd()
setwd("~/GitHub/QuantitativeBiodiversity/Assignments/GeographicalEcology")
```

### Load Packages 
As in previous exercises, we will use the `vegan` package.
We will also use the `BiodiversityR` package, which contains additional diversity estimators and related functions.

```{r, results = 'hide', message = FALSE, warning = FALSE} 
require("vegan")
require("BiodiversityR")
```

We will also use packages developed in R for geographical information systems (GIS). 
These packages will allow us to generate maps of study areas, project diversity data onto environmental layers, and conduct geographical analyses.

```{r}
#install.packages("RgoogleMaps")
library(RgoogleMaps) # overlays of data onto a Google map. #vs. require()???#
```

## 2) LOADING DATA
We will be analyzing data from a survey of shallow ponds found east of Bloomington, Indiana that are scattered throughout Brown County State Park, Yellowwood State Forest, and Hoosier National Forest. 
In the 1940s, Maury Reeves of the Indiana Department of Natural Resources started making ponds with a small dozer in an effort to create a refuge for wildlife, specifically deer and turkey. 
In the summer of 2013, we visited approximately 50 of these "refuge" ponds and recorded their geographic locations using Global Positioning System (GPS). 
In addition, the ponds were sampled for water chemistry, phsyical properties, and bacterial community composition. 

```{r}
#![pond](./data/pond.JPG)
```
## 3) MAKE A MAP
In this section of the exercise, we are going to use R to generate a map using **Google Maps**.
First, we will retrieve a map of Indiana using the `GetMap` function in the package `RgoogleMaps`.
Then, we will generate a map in RStudio that is centered on Brown County, Indiana (39.1 degrees latitude, -86.3 degrees longitude).

### It seems the extent of the maps is a bit large (contains Martinsville); can we zoom in? I'm assumign the zoom arugment?###
```{r}
newmap <- GetMap(center = c(39.1,-86.3), zoom = 10, 
                 destfile = "Brown_Co_map.png", maptype="terrain")
PlotOnStaticMap(newmap, zoom = 10, cex = 2, col='blue') # Plot map in RStudio

Ponds <- read.table(file="BrownCoData/20130801_PondData.csv", head=TRUE, sep=",")
lats <- Ponds[, 3] # latitudes
lons <- Ponds[, 4] # longitudes
PlotOnStaticMap(newmap, lats, lons, cex=1.5, col='red', add = TRUE)
```

***Question 1***: Based on the map that you just generated, what can you conclude about the geographic and topographic distribution of the refuge ponds?

> ***Answer 1***: 

## 4) AGGREGATION

Aggregation is the tendency for organisms to cluster in space or for events to occur within a particular period of time. 
Aggregation also refer to the tendency of a variable to take on a particular value, e.g., mean temperature, modal body size, average abundance.

In this component of the exercise, we will explore aspects of **spatial???** aggregation through graphical exploration **and the calculation of univariate metrics. --> passive?**

### Spatially implicit
Before examining how environmental variables, abundances, and aspects of diversity vary with distance or increasing area, we will examine how they vary within the landscape, regardless of their position in space. **as worded, sort of jargony; will students follow?**

**awkward, abrupt transition here...going from aggretation to exploring distribution of data; perhaps start off with looking at distribution of variables and then investigate spatial aggregation**

### i. Graphically examine environmental data
Before conducting any formal statistical analyses, it's good practice to look at the underlying distribution of your data.
We can ask whether our data appear to be normally distributed, highly skewed, multi-modal, or **even spurious -- not sure I know what this mean in comparison to other distributions mentioned**

Let's begin with a simple graphical exploration of the environmental data collected from the refuge ponds.
We will use **kernel density curves**, which are similar to histograms, but more appropriate for continuous data because they avoid the arbitrary creation of bins.
Kernel density curves attempt to account for uncertainty and reveal the 'probability density' of a variable taking values within a particular range.**this last sentence sounds sort of jargony to me**

**Should we include units of these measures? degrees C, meters = elevation?, feet!?!? = depth, mg/L = DOC
```{r}
EnvData <- Ponds[5:21]
par(mfrow=c(2, 2))

data <- as.numeric(EnvData[, "Temp"])
plot(density(data), col = 'magenta', main = "Temperature")

data <- as.numeric(EnvData[, "Elevation"])
plot(density(data), col = 'DarkCyan', main = "Elevation")

data <- as.numeric(EnvData[, "Depth"])
plot(density(data), col = 'darkorchid', main = "Depth")

data <- as.numeric(EnvData[, "DOC"])
plot(density(data), col = 'darkgoldenrod3', main = "DOC")
```

**Can we describe "bandwidth" on plots? what does this mean?**

***Question 1***:  Describe the distributions of of the environmental variables in the refuge ponds. 

> ***Answer 1***: 

### ii. Graphically examine taxonomic data

Load the site-by-OTU matrix for samples taken from ponds in and around Brown County, Indiana.
**perhaps we need to describe the biological data more above? Bacterial 16S rRNA genes, etc. **

**we've been trying to be consistent and keep variable names lower case**
```{r}
# Site by OTU matrix
OTUs <- read.csv(file="BrownCoData/SiteBySpecies.csv", head=TRUE, sep=",")
otu.names <- names(OTUs)
OTUs <- as.data.frame(OTUs[-1]) # remove first column (site names)
Site.N <- as.vector(rowSums(OTUs)) # no. reads
```

Now, let's construct kernel-density curves for the distribution of abundance among individual OTU's across the Brown County Pond dataset. **I suppose we can just call in "brown county ponds" but above have referred to it as "refuge ponds"; for exercise and publications, maybe we can just pick something; not sure all ponds are actually in brown county**

**More importantly, we need to take some time to teach students about loops. First introduced in alpha exercise, but we never explained. Safe to assume that most student are not familiar with loops or while statements...at least not in R, but I was say more generally, too**
```{r}
ad <- c(0,0)
otu <- 1
while (length(ad) <= 10){
  otu <- sample(1:length(OTUs), 1)
  
  ad <- OTUs[, otu]
  ad = as.vector(t(x = ad))
  ad = ad[ad > 0]

  }
  
plot(density(ad), col = 'magenta', main = otu.names[otu])  
```

As you can see, sampled abundance for a given OTU is often aggregated, revealing many sites where the OTU is relatively rare and many where it is relatively more common.
**In fact, this uneven spatial distribution of abundance is common occurrence in ecological systems.**
This spatially implicit distribution of abundance is often referred to as the species spatial abundance distribution (SSAD).

**I think the concept here needs to be more developed. Maybe we'll talk about SSAD in lecture. I'm afraid most students will have never heard of SSAD. Need to explain. X-axis is not labeled, so not sure what we're looking at. Also, perhaps this is an opportunity to ask a question as opposed to telling them what they're seeing. learning opp**

** Should SSAD be it's own section as opposed to be nestled in with "aggregation"? Seems like theory or concept is a little burried**

***Question 1***: In the site-by-species matrix, each row represents a site and each column represents an OTU. 
If the SSAD is generated by considering all rows for a single column, then what do we obtain when we consider all columns for a given row? 
Have we examined this sort of data structure before? 
If so, elaborate?

**See comment above about introducing a new core concept (i.e., SSAD). Need to teach them what this is first!**

> ***Answer 1***: 


## 5) THE DISTANCE-DECAY RELATIONSHIP
First law of geography...spatial autocorrelation...decrease in similarity with distance...
**Lot more text and development needed*
**Lots of code for them to type out**

```{r, results = 'hide', message = FALSE, warning = FALSE}
#install.packages("simba")
library(simba) #vs. require()???#
```


```{r}
# calculate the similarity (Bray-Curtis) between the plots
struc.dist <- 1 - vegdist(OTUs) 
# calculate geographical distance between plots
coord.dist <- dist(as.matrix(lats,lons))

# transform environmental data to numeric types
temp <- as.numeric(EnvData[, "Temp"])
elev <- as.numeric(EnvData[, "Elevation"])
depth <- as.numeric(EnvData[, "Depth"])
doc <- as.numeric(EnvData[, "DOC"])

# calculate the distance (Euclidean) between the plots regarding environmental variables
env.dist <- 1 - vegdist(cbind(temp, elev, depth, doc), "euclidean")

# transform all distance matrices into list format:
struc.dist.ls <- liste(struc.dist, entry="struc")
env.dist.ls <- liste(env.dist, entry="env")
coord.dist.ls <- liste(coord.dist, entry="dist")

# create a data frame containg plot information, geographical distance, 
# similarity of environment, and similarity of community
df <- data.frame(coord.dist.ls, env.dist.ls[,3], struc.dist.ls[,3])
names(df)[4:5] <- c("env", "struc")
attach(df)
#df <- subset(df, struc != 0)

# plot Distance-Decay relationships with regression lines in red
par(mfrow=c(1, 1))
plot(dist, env)
abline(lm(env ~ dist), col="red4")

par(mfrow=c(1, 1))
plot(dist, struc)
abline(lm(struc ~ dist), col="red4")

# Is the slope significantly different?
res <- diffslope(dist, env, dist, struc)
res
```


## 6) THE SPECIES-AREA RELATIONSHIP
The species-area relationship (SAR) describes the rate at which we discover or accumulate species with increasing area. 
The general relationship of the SAR ... and is of the form S=c*A^z ... which was first predicted by Arrhenius (1921)...wow, that's old.

### Random accmulation of sites

### Accumulation of sites by proximity

```{r}
OTU.obs <- function(x = ""){
  rowSums(x > 0) * 1
  }
```
