---
title: "Week 2 Exercise: Local Diversity"
author: "Z620: Quantitative Biodiversity, Indiana University"
date: "January 23, 2015"
output: pdf_document
---

## Overview 

In this exercise, we will explore aspects of local or site-specific diversity, also known as alpha ($\alpha$) diversity. 
We will begin by introducting one of the primary ecological data structures: the **site-by-species matrix**.
With this data structure, we will quantify two of the fundamental components of ($\alpha$) diversity: **richness** and **evenness**. 
From there, we will then discusss ways to integrate richness and evenness. 
This will include univariate metrics of "diversity" along with an investigation of **species abundance distributions (SAD)**.

## 1) Setup
### Retrieve and Set Your Working Directory

```{r, results = 'hide'}
rm(list=ls()) # removes all variables from your workspace
getwd() # prints the working directory
setwd("~/GitHub/QuantitativeBiodiversity/Assignments/Week2") # sets the working directory
```

### Install Packages

In this excercise, we will rely heavily on a **contributed R package** called `vegan`, which contains tools commonly used in ecological research, including analyses of $\alpha$-diversity. 
Jari Oksanen has created an [excellent tutorial](http://cc.oulu.fi/~jarioksa/opetus/metodi/vegantutor.pdf) that provides an overview of the `vegan` package.

Recall from last week that we can load a package and dependencies using the `require()` function. 
If the package is not found, then we will use the `install.packages()` function followed by `require()`:

```{r, results = 'hide'}
require("vegan")||install.packages("vegan");require("vegan") 
```

## 2) Loading Data

We will start by using the tropical forest dataset from **Barro-Colorado Island (BCI)**.
BCI is a 1,560-hectare island in the center of the Panama Canal that is administered by the [Smithsonian Tropical Research Institution](http://www.stri.si.edu/english/research/facilities/terrestrial/barro_colorado/). 
Conveniently, the `vegan` library contains a version of the BCI dataset. 
The dataset comes from the censusing of tree species in 50 one-hectare plots every several years. 
More information on the BCI forest dataset can be found [here](http://ctfs.arnarb.harvard.edu/webatlas/datasets/bci/). 
Specific details about the BCI dataset associated with the 'vegan' package can be obtained by typing `help(BCI)` in the command line.
Let's load the data:

```{r}
data(BCI)
```

## 3) Exploring The Site-By-Species Matrix

The **site-by-species matrix** is one of the *primary ecological data structures*. 
We will draw inferences from this type of data structure throughout the course. 
The site-by-species matrix contains the abundance, relative abundances, or presence-absence of species (or other taxonomic units) recovered from different sampling sites. 

In `vegan`, the BCI **site-by-species matrix** has the abundances of all 225 tree species in the columns for each of the 50 sites, which are in the rows. 
It is worth noting, that each row could just as easily be a species and each column could be a site (as you will see below with another data set)
Some programs that you encounter will prefer data to be in one format or the other. 
Recall that that transpose function (`t()`) can be used to easily modify the the site-by-species matrix for you. 

Let's verify the structure of the BCI site-by-species matrix using the `dim` function, which was introduced last week:

```{r}
dim(BCI)  # Prints the dimensions of the site (row) by species (column) data frame
```

With the BCI site-by-species matrix now loaded, let's print the abundances of four species found in the first eight sites using the indexing tools that were introduced last week.

```{r, results = 'hide'}
BCI[1:8, 14:17]	# print abundances for sites (rows) 1 to 8 and for species (columns) 14 to 17
```

Here, we can see that Cabbage Bark (*Andira inermis*) is absent from six of the eight sites and is only found as a single individual in two sites (sites 5 and 6). 
On ther other hand, *Apeiba aspera* (locally known as Monkey Comb) is found at all eight sites and is relatively much more abundant than any of the other three species that we indexed in the data frame.

## 4) Species Richness (S)

**Species richness** is the most basic aspect of diversity. 
In fact, it is usually what most people are referring to when they say $\alpha$-diversity.
Estimating species richness for a single sample may seem fairly straight-forward. 
However, there are a few things to consider and pitfalls that we should avoid. 

### Observed Richness

The simplest way to calcuate species richness is to just add up the number of species that have non-zero abundances in a site. 
Let's calculate species richness using this approach for one of the BCI sites. 
Using some of the R basics from last week, let's assign the first row to a variable called "site1":

```{r}
site1 <- BCI[1,] # assign the first row (site) to the variable site1 
dim(site1) # print the dimensions of site1 to the screen 
```

The site could contain up to 225 species (columns). 
It is important to note, however, that site-by-species matrices also account for absences (i.e., zero occurences). 
Let's write a function that estimates observed species richness of a site:

```{r}
S.obs <- function(x){
  rowSums(x>0)*1
  }
```
The use of functions is elemental to programming, in general.
The basic concept behind functions is to define a piece of code that operates on a single or set of variables and data structures.
So, instead of rewriting the code to calculate richness over and over, we just define a function called S.obs that calculate richness of a given vector x. 
We can then type (or "call") the S.obs function whenever we want to calculate richness for a given vector, x.
There is also a function in the `vegan` package called `specnumber`. 
Do we get the same value? Try it! 
In fact, functions in the vegan package are defined and operate the same way that the functions we define and write ourselves.
The benefit of packages such as vegan is that they offer a standard set of functions and statistical tools that we don't have to derive, define, and test, ourselves.

### But How Well Did You Sample Your Site?

Accurate estimates of richness are influenced by sampling effort.
The more individuals that are censused, the more likely you are to encounter new species.
There are a number of indices avaiable that will provide an estimate of how well we've sampled our site. 
One is known as **Good's Coverage (C)**, which is defined as 1 - (**n1**/**N**), where n1 is the number of number of individuals belonging to singleton species, i.e. species encountered only once in a sample, and **N** is the total number of individuals in a sample. 
Let's write a function and estimate Good's Coverage for site1 of BCI:

```{r}
C <- function(x){
  1 - (sum(x==1)/rowSums(x))
  }
```
### Questions about Good's Coverage (C)
Q1. Have the researchers at BCI done a good job of sampling site1?
A1. ...
Q2. What is the range of values that can be generated by C?
A2. ...
Q3. What proportion of "present" taxa in site1 consisted of singletons?
A3. ...

### Richness Estimators

There are few systems on earth that have been better surveyed than the trees at BCI <-actually reveals a shortcoming of Good's coverage (i.e. 1's are not sampling errors). 
(Maybe use Microbe Data for Good's coverage, instead)
So, what happens when we try to characterize the diversity of less-well sampled systems? 
To address this question, we are going to introduce a new data set that is derived from 16S rRNA gene sequences obtained from multiple plots at the [KBS Long-Term Ecological Site](http://lter.kbs.msu.edu/). 
Even though we obtained a fair number of sequences from each sample, soil bacteria are thought to be some of the most divese communities on earth. 
So, how diverse are they?

First, let's load the new data set:

```{r}
soilbac.raw <- read.table("data/soilbac.txt", sep = "\t", header = TRUE)
soilbac <- setNames(data.frame(t(soilbac.raw[,-1])), soilbac.raw[,1]) # transpose and turn column headers into row names so we have a site (rows)-by-"species" (columns) matrix
soilbac1 <- soilbac[1,] # row 1 is a sample taken from an agricultural site
```

How many sequences did we recover from the soil sample?
What is the observed richenss of this sample?
How does coverage compare between the BCI sample (trees) and the KBS soil sample (bacteria)?
###place for students to write out answers###

There are number of techniques that have been developed by biodiversity researchers to estimate (or "extrapolate") species richness based on the coverge of species in a sample (e.g., [Hughes et al. 2001](http://aem.asm.org/content/67/10/4399).
Here, we will hightlight two non-paremetric diversity estimators. 
The first is **Chao1**, which is calculated using observed richness (S.obs) and the number of **singletons** and **doubletons**. 
This estimator tends to be used in communities that are skewed towards low-abundance taxa.
Let's write a function for Chao1:

```{r}
S.chao <- function(x){
  rowSums(x>0)*1 + (sum(x==1)^2)/(2*sum(x==2)) ### should we use which() and length() here? Is it confusing to switch to something new for ACE?
  }
```

Another commonly used estimator is **ACE**, which stands for abundance-based coverage estimator. 
While Chao1 makes inferences based on the number of singletons and doubletons, ACE implements a threshold to look at the abundance of other *rare* species. 
By convention, ACE defines rare species as taxa that have 10 or fewer individuals.
Consequently, whether one uses the ACE estimator depends on whether one's samples tend to have many species of few individuals.
If so, the ACE estimator may ignore the majority of sampled species.
Now we will write a function for the ACE estimator:

```{r}
S.ace <- function(x = "", thresh = 10){ 
  x <- x[x>0]                           # excludes zero-abundance taxa
  S.abund <- length(which(x > thresh))  # richness of abundant taxa
  S.rare  <- length(which(x <= thresh)) # richness of rare taxa
  singlt  <- length(which(x == 1))      # number of singleton taxa
  N.rare  <- sum(x[which(x <= thresh)]) # abundance of rare individuals
  C.ace   <- 1 - (singlt / N.rare)      # coverage (prop non-singlt rare inds)
  i       <- c(1:thresh)                # threshold abundance range
  count   <- function(i, y){            # counter to go through i range
    length(y[y == i])
  }
  a.1     <- sapply(i, count, x)        # number of individuals in richness i richness classes
  f.1     <- (i * (i - 1)) * a.1        # k(k-1)kf sensu Gotelli
  G.ace   <- (S.rare/C.ace)*(sum(f.1)/(N.rare*(N.rare-1)))
  S.ace   <- S.abund + (S.rare/C.ace) + (singlt/C.ace) * max(G.ace,0) 
  return(S.ace)
}
```
### This is pretty hairy I guess. -Jay
### Can simplify your code, i.e. make it cleaner? -Ken

### Opportunity to teach about loops and alternatives in R (sapply). -Jay
### sapply is nice if the task is relatively simple. -Ken
### This is a bad example for teaching loops. -Ken
### My opinion: Loops are too rudimentary a skill and this function is too complicated to stop inside of and say, "While you're thinking about estimating richness we're going to teach you a new control structure" -Ken

### Or we can just drop it and stick with Chao1; I'm guessing -- conservatively -- it may take 15 minutes to get this transcribed from handout and running. -Jay
### And the question is: what will they take away? -Jay
### Agreed. -Ken
### Maybe do Chao1 and Chao2 in class. Then, for homework, give a short explanation of ACE and then say, "Using the example provided, code up the ACE estimator and estimate species richness for sample x" -Ken

### One idea I had is that we could put "bits" of stuff that we do in "Supplementary" files. Even if students don't use it right away, it's there for them, but perhaps more importantly us! -Jay
### If the students don't use it, then they should see it; avoiding clutter. -Ken
### But yes, might do well in the instructor materials...or better as homework -Ken

How does observed richness compare to estimated richness for soilbac1?

_Notes_
+ `estimateR` is a function in the `vegan` pakcage 
+ It will spit out observed richness, along with Chao1, ACE, and their associated confidence intervals. 
+ You can look more into the code in R packages using commands like this: vegan:::estimateR.default
+ Try it out!
## estimateR gives Chao estimates that are very close, but not identical to S.chao. 
## estimate R give ACE estimates thar are significantly off, perhaps due to their error. 
  
## Rarefaction

Often researchers want to compare the richness of two or more samples. 
For example, is the richness of invertebrate species in a contaminated stream different from an undisturbed habitat located upstream? 
Ideally, we would sample all of our sites with equal effort, but for many reasons this often does not happen.
One way to reduce this bias is to **rarify** our samples down to a "lowest common denominator"" of sampling effort. 

Let's looks at the richness of soil bacteria for all 12 of the KBS sites, where T1 = agriculture, T7 = grassland, DF = deciduous fores, and CF = coniferous forest. 

```{r}
soilbac.S <- S.obs(soilbac)           # observed richness
min.N <- min(rowSums(soilbac))        # sample with fewest sequences
S.rarefy <- rarefy(soilbac, min.N)    # richness when rarefied to sample with fewest sequences
rarecurve(soilbac, step = 20, sample = min.N, col = "blue", cex = 0.6, las=1) # vertical line = min.N; horizontal lines S at min.N
abline(0, 1, col = 'red')             # 1:1 line for reference
text(1500, 1500, "1:1", pos = 2, col = 'red')
```
### I think there's probably some math to explain here -Jay (?)
### Agreed; I'd want students to know just what 'rarefy' does and why.

### what about errors associated with estimates? -Jay (?)
### Would be more rigorous with errors or CIs. -Ken

### how about species accumulation curves using specaccum()? -Jay(?) 
### Redundant with rarefy? -Jay(?)
### Not really: Species accumulation curves show the rate at which new species are found within a community and can be extrapolated to provide an estimate of species richness. -Ken
### [Clarification](http://www.vsni.co.uk/software/genstat/htmlhelp/ecology/AccumulationCurve.htm); A big literature on SACs and collector's curve. -Ken

##5) Species evenness
There is more to ($\alpha$) diversity than just the number of species in a plot. 
Specifically, it is important to consider how abundance varies among species, that is, **species evenness**.
Many important biodiversity issues -- species coexistence, community stability, the detection of rare taxa, the observation of compositional change and turnover due to disturbance, invasion, and random population fluctuations -- relate to variation in the abundance of taxa. 

### Visualizing Evenness: The Rank Abundance Curve (RAC)
One of the most common ways to visualize evenness is to rank the species from most abundant to least abundant without respect to species labels (and hence no worries about 'ties' in abundance). This **rank-abundance curve (RAC)** is also referred to as a rank-abundance distribution and as a Whittaker plot. 

Let's construct an RAC using site1 of the BCI data set. 
First, we will remove species that had zero occurences in site1. 

```{r}
site1.ab <- site1[site1>0] # Remove all zeros from Site1 and reassign the new vector
```

```{r}
site1.ab.ranked <- site1.ab[order(site1, decreasing = TRUE)] # Use `order` function and "decreasing" argument to rank taxa by abundance
ranks <- as.vector(seq(1, length(site1)))             # Create a sequence of S ranks
plot(ranks, log(site1.ab.ranked), type = 'p', las = 1,      # Plot the RAC
     xlab = "Rank", ylab = "log(abundance)")           # "type = 'p'" = points; "las = 1" labels are horizontal
```

It is clear from looking at the RAC for site1 that abundance among species is unequally distributed. 
This is true even when abundances are log-transformed (note the plotting options above), which deemphasizes the influence of the few most abundant species.
That is, we don't want our interpretation of the overall pattern to be entirely driven by a few species.
This sort of uneven distribution of abundance among species is one of the most ubiquitous patterns in ecology and has provoked a long history of study, theories, and explanations [McGill et al. 2007](http://www.ncbi.nlm.nih.gov/pubmed/17845298).

### Species Evenness
Species evenness is generally defined as the similarity in abundance among species. 
If **N**/**S** equals average species abundance of a sample, you can think of species evenness as being related to the samples variance.
Researchers have attempted to identify important features that evenness measures should strive for:

1. We want to be able to intuit the values of our metric. 
2. We want our metric to be bounded between a minimum evenness (i.e. 0.0) and a maximum evenness (i.e. 1.0). 
3. We also want intermediate values of our metric (e.g. 0.5) to represent an intuitive intermediate evenness.
4. We do not want evenness among species to simply reflect the number of species (i.e. *S*); otherwise, comparing evenness among sites with different numbers of species really becomes a comparison of species richness; not evenness *per se*. 
5. Adding a singleton species should decrease our the value of evenness.

In short, desirable evenness metrics are bounded, unbiased, and have intuitive values.
Here, we will introduce two metrics of evenness that meet the above criteria: Simpson's evenness and Smith and Wilson's evenness index 

### Simpson's evenness, $E_{1/D}$

Simpson's evenness metric essentially reflects the sample variance of the SAD, and is calculated as ${E_{1/D} = } \frac{1}{S} \sum \frac{N(N-1)} {n_i(n_i-1)}$, where *S* is species richness, *N* is total abundance, *i* is the *i*th species. 

Once, again vegan has no function for Simpson's evenness but does calculate Simpson's diversity, i.e., the probability that the next sampled individual belongs to a different species:

```{r}
D <- diversity(site1.ab, "simp") # find Simpson's diversity (D)
S <- specnumber(site1.ab)
E <- (1/D)/S # find Simpson's evenness (E)
E
```

Returning to the BCI tropical forest dataset, we can see that Simpson's evenness reveals that species evenness for site1 is quite low, i.e., < 0.1. However, Simpson's evenness has been criticized for being biased towards the most abundant taxa, i.e., revealing differences in abundances among the first few ranks and is less influenced by the many similarly abundant but rarer species.
Let's examine the value of evenness for Site1 using a different and evenness metric that is less biased in this way, i.e., $E_{var}$.

### Smith and Wilson's evenness index, $E_{var}$

Smith and Wilson (1996) reviewed metrics of species evenness. 
Some of their conclusions about the desired characteristics of species evenness are reflected in this assignment. 
They derived a robust measure of species evenness called ($E_{var}$). 
This metric is a measure of the sample variance of logarithmic abundances, standardized to take values between 0 (no evenness) and 1 (perfect evenness).
Abundances are transformed to their natural logarithms to decrease bias towards the most abundant species, that is, the potential for a metric's value to be influenced more by large numbers than small ones.
$E_{var}$, like all desireable measures of evennness, is independent of richness *S*.
The metric is calculated as: ${E_{var} = 1 - } \frac{2} {\pi \cdot \arctan( \sum\limits_{i=1}^{i=S} ln(n_i) - \sum\limits_{j=1}^{j=S} ln(n_j)/S)}$.

While seeminly more involved to calcualte, $E_{var}$ simply reduces to finding the sample variance of the log-transformed species abundances and then standardizing it to take values between 0 and 1 by using elementary trigonometry.

```{r}
X <- var(log(site1.ab))
Evar <- 1 - (2/pi)*atan(X) # these operations make the value of Evar range between 0 and 1
Evar # print Evar for Site1
```
$E_{var}$ uses the arctangent, which varies between -$\pi$/2 and $\pi$/2 and without being periodic like waves of the sine and cosine functions. Multiplying the arctangent by 2/pi forces the result to take values between 0 and 1. Finally, subtracting this from one allows low evenness to be associated with values near 0 and high evenness to be associated with values near 1. We can confirm this with a more explicit R chunk:

```{r}
P <- log(log(site1.ab)) # log-tranform the abundances of the RAC and assign them to a vector P
AvgAb <- mean(P) # find the average of the log abundances
X <- 0 # assign zero to variable X
Evar <- 0 # declare a scalar varible Evar

for (x in P) { # making use of a 'for' loop. for loops are an elementary control structure in all programming languages.
  X = X + (x - AvgAb)^2 / (S - 1)
}

Evar = 1 - (2/pi)*atan(X) # these operations make the value of Evar range between 0 and 1
Evar # print Evar for Site1
```

As you can see, the value of $E_{var}$ suggests an intermediate value of species evenness. 
Likewise, you can see that different evenness metrics, even when bound between 0 and 1 still return very different values. 
This is not because each is inaccurate but because each metric emphasizes different aspects of the RAC. 
Where Simpson's evenness focuses more greatly on dominant species, $E_{var}$ gives nearly equal weight to dominant, intermediately abundant, and rare species.

* not sure if we should use BCI or bacteria data for above. could choose one and then ask them to do other for homework and compare!

### You're going from observed values (e.g. richness), to estimated values, and then back to observed values. -Ken
### Are you sure that you also want to go from BCI, to microbes, and back to BCI? -Ken
### Could be more linear by just going from observed to estimated and/or from BCI (or microbes) to microbes (or BCI). -Ken

k <- sample(nrow(BCI), 1) ## take a random site from BCI
### I would suggest not doing this because you might get a pretty crappy site. -Ken
### Best to have control over it, especially since students homework won't end up agreeing. -Ken
### Maybe just use the original site chosen from BCI (site1), which I had already checked for having a decent number of species. -Ken
fisher <- fisherfit(BCI[k,]) # using an unambiguous naming convention

```{r}
N <- sum(site1.ab) # find the number of individuals at site1 and assign it to a variable N
cat('There are', N, 'individuals among', S, 'species at Site1') 
```


##6) Integrating richness and evenness: "diversity metrics"
Metrics developed from information theory. Commonly used. Rarely understood.

##7) Integrating richness and evenness: Species Abundance Distributions (SAD)

### Be careful not to confuse "diversity metrics" with the SAD as so-called 'integrations' of richness and evenness. -Ken


## Species Diversity
You saw in the derivation of Simpson's evenness that we called the R diversity function, and then specified "simp". 
In effect, in estimating Simpson's evenness we first estimated Simpson's diversity (*D*). Species diversity is commonly defined as a relationship between species richness and species evenness, often of a form similar to ${Diversity = } \frac{Evenness} {Richness}$.
Here, we will estimate popular indices of diversity using our own derivation and then check this against vegan's estimates. 

### For the purpose of..." is the learning objective to just check vegan to make sure it's right? -Jay
### No. The purpose is to instruct students in the ways of understanding that packages like vegan don't do anything special. -Ken
### Not only is there no magic, but even us normal folk can code up the functions we see in papers, books, and usually leave to super-geeks to develop. -Ken 
### It's about being empowered and non-phobic through comprehending. -Ken

### Shannon's diversity (or entropy)
Shannon's diversity metric is really just Shannon's information entropy, a measure of uncertainty. 
This metric is calculated as ${H' = -}  \sum p_i ln(p_i)$.
Let's calculate Shannon's diversity for Site1 and then compare it to vegan's estimate:

```{r}
RAC <- site1.ab.ranked
H <- 0
for (sp in RAC){
  p = sp / sum(RAC)
  H = H - p*log(p) 
}
H

diversity(RAC, index="shannon")
```

### Are they the same? -Mario
### Unless you have a Ph.D. in Statistical physics or Information theory and can actually geek out on various forms of entropy, Yes. -Ken
### [Claude Shannon's original paper](http://cm.bell-labs.com/cm/ms/what/shannonday/shannon1948.pdf) -Ken

### Simpson's diversity (or dominance)
Simpsons diversity is a straightforward metric and is calculated as ${D = } \sum p_i^2$ where $p_i$ is the proportion of individuals found in the *i*th species. 
Simpson's index is often expressed as 1/*D* or 1-*D*, so that diversity naturally increases with 1/*D*. 
Let's calculate Simpson's diversity for Site1 and then compare it to vegan's estimate:

```{r}
D <- 0.0
N <- sum(RAC)
for (ni in RAC){
  D = D + (ni*ni)/(N*N)
}

invD <- 1/D # using the 1/D variation
invD

invD <- diversity(RAC, "inv") # using vegan
invD

D <- 1 - D # using the 1 - D variation
D

D <- diversity(RAC, "simp") # using vegan
D
```

### Fisher's $\boldsymbol\alpha$
R.A. Fisher (1943) derived one of the first and most successful models for how abundance varies among species, i.e., the log-series distribution. 
This model has only a single fitted parameter, i.e., $\alpha$, which is implicitly defined by the log-series model: ${S =} \alpha * ln(1 + n/\alpha)$. 
However, this does not tell us exactly how to estimate $\alpha$, and this is because $\alpha$ is a fitted parameter. 
Later, we will explore the log-series. 
For now, let's concentrate on the Fisher's $\alpha$, which has often been used as a diversity metric and is the root of 'alpha' diversity. 
Fisher's $\alpha$, according to the authors of the vegan package is asymptotically similar to inverse Simpson's. 
Let's compare using Site1.

```{r}
invD <- diversity(RAC, "inv")
invD
Fisher <- fisher.alpha(BCI)
Fisher
```

As we can see, the two measurements are somewhat similar. 
They would converge if our community was much greater in total abundance and richness. 
However, discussion of Fisher's $\alpha$ introduces a new concept, that is, of estimating diversity instead of just calculating a diversity metric. 
The difference being that an estimate of diversity implicitly or explicitly accounts for samplign error, that is, the fact that when samplign most ecological communities that we are not observing every single individual.

## Species abundance models
Recall that the species rank-abundance curve RAC is simply the abundance of species ranked from most-to-least abundant.
The RAC is a simple data structure that is both a vector of abundances and a row in the site-by-species matrix (minus the zeros, i.e., absences). 
Despite it's simplicity the RAC contains a lot of information.
For example, the sum of the abundances in the RAC is commonly referred to as the observed **total abundance** and denoted as **N** while the length of the RAC (the number of ranks) is simply the observed **species richness**, denoted as **S**. 
The RAC also contains the information needed to estimate richness and calculate evenness and diversity.
Moreover, the RAC represents a type of species-abundance distribution (SAD), the form of which has been intensively studied and modeled.
Species abundance models attempt to predict the form of the SAD (e.g. as an RAC) according to mechanisms and processes that underpin general ecological and biodiversity theory.

Let's use the radfit() function in the vegan package to fit the predictions of various species abundance models to the RAC of site1 in BCI 
```{r}
RACresults <- radfit(RAC)
RACresults
```

___

## Homework

1) As stated by Magurran (2004) the  ${D = } \sum p_i^2$ derivation of Simpson's D really only applies to communities of infinite size. #For anything but an infinitely large community, Simpson's index is calculated as ${D = } \sum \frac{n_i(n_i -1)} {N(N-1)}$. 
Calculate Simpson's D, 1 - D, and Simpson's inverse (i.e. 1/D) for Site1.

2) ...

3) use knitr to create a pdf, push it to GitHub, and create a pull request

