---
title: "Week 2 Exercise: Local Diversity"
author: "Z620: Quantitative Biodiversity, Indiana University"
date: "January 23, 2015"
output: pdf_document
---

## Overview 

In this exercise, we will explore aspects of local or site-specific diversity, also known as alpha ($\alpha$) diversity. 
We will begin by introducting one of the primary ecological data structures: the **site-by-species matrix**.
From this data structure, you will learn how to quantify two of the fundamental components of ($\alpha$) diversity: **richness** and **evenness**. 
We will then discusss ways to integrate richness and evenness. 
This will include univariate metrics of "diversity" along with and investigation of **species abundance distributions (SAD)**.

## 1) Setup
### Retrieve and Set Your Working Directory

```{r, results = 'hide'}
rm(list=ls()) # removes all variables from your workspace
getwd() # prints the working directory
setwd("~/GitHub/QuantitativeBiodiversity/Assignments/Week2") # sets the working directory
```

### Install Packages

In this excercise, we will rely heavily on a contributed R package called `vegan`, which contains tools commonly used in ecological research, including analyses of $\alpha$-diversity. 
Jari Oksanen has created an [excellent tutorial](http://cc.oulu.fi/~jarioksa/opetus/metodi/vegantutor.pdf) that provides an overview of the `vegan` package. 

Recall from last week that we can load a package and dependencies using the `require()` function. 
If the package is not found, then we will use the `install.packages()` function followed by `require()`:

```{r, results = 'hide'}
require("vegan")||install.packages("vegan");require("vegan") 
```

## 2) Loading Data

We will be using the tropical forest dataset from **Barro-Colorado Island (BCI)**.
BCI is a 1,560-hectare island in the center of the Panama Canal that is administered by the [Smithsonian Tropical Research Institution](http://www.stri.si.edu/english/research/facilities/terrestrial/barro_colorado/). 
Conveniently, the `vegan` library contains a version of the BCI dataset. 
The dataset comes from the censusing of tree species in 50 one-hectare plots every several years. 
More information on the BCI forest dataset can be found [here](http://ctfs.arnarb.harvard.edu/webatlas/datasets/bci/). 
Specific details about the BCI dataset associated with the 'vegan' package can be obtained by typing `help(BCI)` in the command line.
Let's load the data:

```{r}
data(BCI)
```

## 3) Exploring The Site-By-Species Matrix

The **site-by-species matrix** is one of the primary ecological data structures. 
We will draw inferences from this type of data structure throughout the course. 
The site-by-species matrix contains the abundance, relative abundances, or presence-absence of species (or other taxonomic units) recovered from different sampling sites. 

In `vegan`, the BCI **site-by-species matrix** has the abundances of all 225 tree species in the columns for each of the 50 sites, which are in the rows. 
It is worth noting, that each row could just as easily be a species and each column could be a site. 
Some programs that you encounter will prefer data to be in one format or the other. 
Recall that that transpose function (`t()`) can be used to easily modify the the site-by-species matrix for you. 

Let's verify the structure of the BCI site-by-species matrix using the `dim` function, which was introduced last week:

```{r}
dim(BCI)  # Prints the dimensions of the site (row) by species (column) data frame
```

With the BCI site-by-species matrix now loaded, let's print the abundances of four species found in the first eight sites using the indexing tools that were introduced last week.

```{r, results = 'hide'}
BCI[1:8, 14:17]	# print abundances for sites (rows) 1 to 8 and for species (columns) 14 to 17
```

Here, we can see that Cabbage Bark (*Andira inermis*) is absent from six of the eight sites and is only found as a single individual in two sites (sites 5 and 6). 
On ther other hand, *Apeiba aspera* (locally known as Monkey Comb) is found at all eight sites and is relatively much more abundant than any of the other three species that we indexed in the data frame.

## 4) Species Richness (S)  

**Species richness** is the most basic aspect of diversity. 
In fact, it is usually what we refer to as $\alpha$-diversity.
Estimating species richness for a single sample may seem fairly straight-forward. 
However, there are a few things to consider and pitfalls that we should avoid. 

### Observed Richness

The simplest way to calcuate species richness is to just add up the number of species that have non-zero abundances in a site. 
Let's calculate species richness using this approach for one of the BCI sites. 
Using some of the basics from last week, let's assign the first row to a variable called "site1":

```{r}
site1 <- BCI[1,] # assign the first row (site) to the variable site1 
dim(site1) # print the dimensions of site1 to the screen 
```

The site could contain up to 225 species (columns). 
It is important to note, however, that site-by-species matrices also account for absences (i.e., zero occurences). 
Let's write a function that will allow us to estimate observed species richness of a site:

```{r}
S_obs <- function(x){
  rowSums(x>0)*1
  }
```

There is also a function in the `vegan` package called `specnumber` that should provide the same value. Try it. 

### But How Well Did You Sample Your Site?

Accurate estimates of richness are invariably influneced by sampling effort.
The more individuals that are censused, the more likely you are to encounter new species. 
There are a number of indices avaiable that will provide an estimate of how well we've sampled our site. 
One is known as **Good's Coverage (C)**, which is defined as 1 - (n1/N), where n1 is the number of **singletons** (i.e., species that are encountered only once in a sample) and **N** is the total number of individuals in a sample.
Let's write a function and estimate Good's Coverage for site1 of BCI:

```{r}
goods <- function(x){
  1 - (sum(x==1)/rowSums(x))
  }
```

Have the researchers at BCI done a good job of sampling site1?
What is the range of values that can be generated by Good's Coverage?
What proportion of "present" taxa in site1 consisted of singletons?

### Richness Estimators

There are few systems on earth that have been better surveyed than the trees at BCI. 
What happens when we are trying to characterize diversity of less-well sampled systems? 
To address this question, we are going to load a new data set that is derived from 16S rRNA gene sequences obtained from multiple plots at the [KBS Long-Term Ecological Site](http://lter.kbs.msu.edu/). 
Even though we obtained a fair number of sequences, soil bacteri are thought to be some of the most divese communities on earth. 
So, how diverse are they?

First, let's load the new data set:

```{r}
soilbac <- read.table("soilbac.txt", sep = "\t", header = TRUE)
soilbac_t <- setNames(data.frame(t(soilbac[,-1])), soilbac[,1]) # transpose and turn column headers into row names so we have a site (rows)-by-"species" (columns) matrix
soilbac1 <- soilbac_t[1,] # row 1 = one site from agricultural soil
```

How many sequences did we recover from the soil sample?
What is the observed richenss of this sample?
How does coverage compare between the BCI sample (trees) and the KBS soil sample (bacteria)?

There are number of techniques that have been developed by biodiversity researchers to estimate (or "extrapolate") species richness based on the coverge of species in a sample (e.g., [Hughes et al. 2001](http://aem.asm.org/content/67/10/4399). 

Here, we will hightlight two non-paremetric diversity estimators. 
The first is **Chao1**, which is calculated using observed richness (S_obs) and the number of singletons and **doubletons**. 
This estimator tends to be used in systems that are skewed towards low-abundance classes.
Let's write a function for Chao1:

```{r}
S_chao <- function(x){
  rowSums(x>0)*1 + (sum(x==1)^2)/(2*sum(x==2)) ## need to square singletons; 2x doubletons
  }
```

Another commonly used estimator is **ACE**, which stands for abundance-based coverage estimator. 
While Chao1 makes inferences based on the number of singletons and doubletons, ACE implements a threshold to look at the abundance of other *rare* species. 
By convention, ACE defines rare species as taxa that have 10 or fewer individuals. 
The ACE estimator requires the following:

```{r}
x <- soilbac1 # Change this according to what sample you want to analyze
thresh<-10 # Rarity threshold (# of individuals)
S_abund <- rowSums(x>thresh)*1 # Species richness when abundance exceeds threshold
S_rare <- rowSums((x <=thresh & x>0)*1) # Species richness when abundances is less than thresh, but greater than zero; note S_abund + S_rare = S_total
singlt <- sum(x==1) # number of singletons
N_rare <- sum((x<=thresh)*x) # Abundance (# of individuals) across rare species
C_ace <- 1 - sum(x==1)/N_rare # ACE coverage; 1 minus number of singleton species divided by abundance of rare species; i.e., the proportion of all individuals in rare species that are not singletons.

f=c() # create empty vector of richness classes
f[1] = 1 # starting richness class of f (i.e. singleton)
for (i in 1:10){ # counter for richness classes
  f[i]=sum(x==i)*(i*i-1) # loop 
  }

G_ace <- (S_rare/C_ace)*(sum(f)/(N_rare*(N_rare-1)))
#G_ace <- ifelse((S_rare/C_ace)*(sum(f)/(N_rare*(N_rare-1))) > 0, G_ace, 0)  

S_ace <- function(x){
  S_abund + (S_rare/C_ace) + (singlt/C_ace) * G_ace
}
```  

So, what do you think about the richness estimates for bacteria in the KBS soilbac1 sample?

_Notes_
+ `estimateR` is a function in the `vegan` pakcage 
+ It will spit out observed richness, along with Chao1, ACE, and their associated confidence intervals. 
+ You can look more into the code in R packages using commands like this: vegan:::estimateR.default
+ Try it out!
  
## Rarefaction

Often researchers want to compare the richness of two or more samples. 
For example, is the richness of invertebrate species in a contaminated stream different from an undisturbed habitat located upstream? 
Ideally, we would sample all of our sites with equal effort, but for many reasons this often does not happen.
One way to reduce this bias is to **rarify** our samples down to a "lowest common demoniator"" of sampling effort. 

```{r} 

## Graphically, the black horizontal lines are a bit distracting, unnecessary {JTL}
 
BCI_S <- S_obs(BCI) # richness for all 50 BCI sites
raremax <- min(rowSums(BCI)) # sample with lowest number of individuals
S_rarefy <- rarefy(BCI, raremax) # use `rarefy` function in R
plot(BCI_S, S_rarefy, las = 1,
     xlab = "Observed No. of Species", ylab = "Rarefied No. of Species")
abline(0, 1)                  # Adds line with interercep = 0, slope = 1

##what are we trying to demonstrate with plot above? the fact that rarefaction tends to reduce obsered richness?##

rarecurve(BCI, step = 20, sample = raremax, col = "blue", cex = 0.6, las=1)
#we need to provide more detail (or I need to learn more); can we remove the black horizontal lines to make less busy? what are the numbers in boxes? 
abline(0, 1, col = 'red')
text(105, 105, "1:1", pos = 4, col = 'red')
###Example: http://www.jennajacobs.org/R/rarefaction.html
```

##5) Species evenness
There is more to ($\alpha$) diversity than just the number of species in a plot. 
As you will soon see, we can learn a lot more about biodiversity of local sites by considering the total number of individuals recorded at Site1.

```{r}
N <- sum(Site1) # find the number of individuals at site1 and assign it to a variable N
cat('There are', N, 'individuals among', S, 'species at Site1') 
```

Now we know the recorded species richness (*S*) and recorded total abundance (usually denoted as *N*) of Site1 in the BCI site-by-species matrix. 
*N* and *S* are two of the most common pieces of data we obtain when sampling ecological communities. 
And from this, we can gain insight into some other important attributes of biodiversity.

---> Pick a couple of the evenness metrics below. Not sure we need 3-4 pages of exercise devoted to this. Can we focus on the **greatest hits**?







##6) Integrating richness and evenness: "diversity metrics"
Stuff developed from information theory. Useful and commonly used. Go through some examples. Upshot, however, is that there's loss of information. Segue into SADs

##7) Integrating richness and evenness: Species Abundance Distributions (SAD)

The Species Abundance Distribution

A next natural step in exploring aspects of local diversity is to ask how abundance varies among species. 
One of the most common ways to visually explore this relationship is to rank the species from greatest (most abudnant) to least (most rare), that is, as a rank-abundance curve (RAC). 

To make an RAC for site1, first, we will remove zeros (i.e., species that, while in the BCI site-by-species matrix, were not found at Site1.)

```{r}
site1 <- site1[!site1 %in% c(0)] # Remove all zeros from Site1 and reassign the new vector
site1b <- subset(site1, select = site1 > 0) # Better use of R commands {MEM}; [whichever one we use, we need to explain better]
```

Now, we can plot the RAC, accounting only for the species that were recorded at Site1.

```{r}
site1.rank <- site1[order(site1, decreasing = TRUE)]  # Use `order` function and "decreasing" argument to rank taxa by abundance
RAC <- as.vector(site1, mode = 'numeric')             # Convert Site1 abundances into a vector ### THIS SEEMS UNECESSARY; DELETE? {JTL}
ranks <- as.vector(seq(1, length(site1)))             # Create a sequence of S ranks
plot(ranks, log(site1.rank), type = 'p', las = 1,            # Plot the RAC
     xlab = "Rank", ylab = "log(abundance)")          # "type = 'p'" = points; "las = 1" labels are horizontal; axes labes = "xlab" and "ylab"
```

It is clear from looking at the RAC for site1 that the abundance of individuals is unequally distrbuted. This is true even when abundances are log-transformed (note the plotting options above). 
This sort of uneven distribution of abundance among species is one of the most ubiquitous patterns in ecology and has provoked a long history of study, theories, and explanations [McGill et al. 2007](http://www.ncbi.nlm.nih.gov/pubmed/17845298).

You may wonder whether all sites (rows) in the BCI site-by-species matrix also have highly uneven RACs. 
But, before we try to answer this question, we are going to explore some other aspects of local diversity at Site1. 
Specifically, we are going to quantify **evenness**, **dominance**, and **rarity** among the species recorded at site1.


**--> We hve discussed introducing different models of the SAD, for example, the log series, lognormal, geometric as dealt with in Chapter 2 of Magurran (2004) and chapter 9 by McGill of Biological Diversity by Magurran and McGill (2011). We discussed fitting an SAD to these different models and using AIC to assess which model is best fit by our data. We need to discuss whether this is something we want to do. I think it has value. Seems like this would be the place to do it. 


## Species Evenness

Species evenness is generally defined as the similarity in abundance among species. 
Take for instance, the RAC for Site1. 
While there are many species with similar but low abundances, the few species with relatively high abundances actually drives the pattern to be uneven and, hence, the slope of the RAC to be steep. 
In fact, one way of quantifying evenness is simply to find the slope of regression through the points in RAC. 
This is referred to as the NHC evenness index (Nee et al. 1992).

```{r}
fit  <- lm(log(RAC) ~ ranks)                 # Simple Linear Regression
summary(fit)
plot(ranks, log(RAC), type = 'p', las = 1,
     xlab = "Rank", ylab = "log(abundance)") 
abline(fit, col = 'red')                     # Add Regression Line
```

Note, that the slope of this relationship must be either 0.0 or negative because the RAC is constrained to between perfect evenness (slope of 0) and infinite unevenness (negative infinity). 
Also, notice that while a slope of 0 implies a perfectly even distribution of abundance among spcies, a slope of -0.035 really doesn't convey much. 
Likewise, the number of ranks (i.e. species) influences the slope by 'stretching' out the relationship. 
Imagine what the slope would be for *S* = 100,000 ranks!
** Section above (following regression code) could be convey major points more clearly. 

### Desired properties of evenness ** How about starting hear? Does anyone really use the NHC? If this isnt' a good metric, do we really need to spend lots of time covering it? Perhaps we can pick 2 measures of evenness. A commmon one like Pielou's, which has it's issues and EVAR, one that is "good". Exercise could be to compare and contrast {JTL}**


Shortcomings of the NHC index reveal important considerations when quantifying species evenness. 

1. We want to be able to intuit the values of our metric. 
2. We want our metric to be bounded between a minimum (e.g. 0.0) and a maximum (e.g. 1.0). 
3. We do not want evenness among species to simply reflect the number of species (i.e. *S*); otherwise, comparing evenness among sites with different numbers of species really becomes a comparison of species richness; not evenness *per se*. 

We also don't want species evenness to simply reflect differences among the most abundant or among the most rare species. 
In short, we desire an intuitive, bounded, and largely unbiased metric of species evenness.

Seveal well-known metrics of evenness have been derived (see Smith and Wilson 1996) that range between 0 (no evenness) and 1 (perfect evenness), are largely independent of the number of species (i.e. *S*), and are more or less unbiased towards either very abundant species or very rare species. 
Here, we will introduce three: Simpson's evenness, Smith and Wilson's evenness index, and Shannon's evenness. **Not the order you present them; also, that is a long sentence  {MEM}**

### Shannon's (or Pielou's) evenness, J'
Shannon's evenness metric is derived from Shannon's information entropy, a measure of uncertainty, and is standardized by *S* to remove the bias of *S* on the statistic. 
Shannon's evenness (*J'*) is calculated as: ${J' = }  \frac{H'} {ln(S)}$, where *H'* is Shannon's information entropy and where $p_i$ is the relative abundance of the *i*th species. 
Shannon's evenness was derived by Pielou (1969, 1975) and is sometimes referred to as Pielou's evenness.

The vegan package does not have functions to calculate Shannon's evenness but does have a function to calculate Shannon's entropy, sometimes referred to as Shannon's diversity, the Shannon-Weiner diversity index, among others.

```{r}
H <- diversity(Site1, index="shannon") # find Shannon's entropy (H')
J <- H/log(S) # find Shannon's evenness (J')
J
```

You can see that Shannon's evenness is quite high for Site1. 
Yet, this doesn't intuitively reflect what we saw in the RAC. In fact, because of the way Shannon's is calculated, values for ecological communities are commonly greater than 0.5. 
Consequently, an intermediate value of *J'* does not necessarily mean an intermediate evenness. Compare this to another evenness metric, i.e., Simpson's evenness.


**What is the takehome? from the Shannon/Pielou's section? Confusing because we're jumping into diversity metrics (Shannon-Weiner) {JTL}*

### Simpson's evenness, $E_{1/D}$

Simpson's evenness metric essentially reflects the sample variance of the SAD, and is calculated as ${E_{1/D} = } \frac{1}{S} \sum \frac{N(N-1)} {n_i(n_i-1)}$, where *S* is species richness, *N* is total abundance, *i* is the *i*th species. 

Once, again vegan has no function for Simpson's evenness but does calculate Simpson's diversity, i.e., the probability that the next sampled individual belongs to a different species:

```{r}
D <- diversity(Site1, "simp") # find Simpson's diversity (D)
E <- (1/D)/S # find Simpson's evenness (E)
E
```

Here, we can see that Simpson's evenness reveals that species evenness for Site1 is quite low, i.e., > 0.1. ** should this be <0.1???**
However, Simpson's evenness has also been criticized for being biased towards the most abundant taxa, i.e., revealing differences in abundances among the first few ranks and is less influenced by the many similarly abundant but rarer species. 
Let's examine the value of evenness for Site1 using a different and purportedly less-biased evenness metric, $E_{var}$.

### Smith and Wilson's evenness index, $E_{var}$

Smith and Wilson (1996) reviewed metrics of species evenness. 
Some of their conclusions about the desired characteristics of species evenness are reflected in this assignment. 
They derived a robust measure of species evenness called ($E_{var}$). 
This metric represents the sample variance of logarithmic abundances. 
Abundances are **natural** log-transformed **we need to be clear about what the base log here and elsewhere {JTL}* to decrease bias towards very rare or very abundant species. 
$E_{var}$ is independent of *S* and is standardized to take values between 0 (no evenness) and 1 (perfect evenness). 
The metric is calculated as: ${E_{var} = 1 - } \frac{2} {\pi \cdot \arctan( \sum\limits_{i=1}^{i=S} ln(n_i) - \sum\limits_{j=1}^{j=S} ln(n_j)/S)}$.

While considerably more involved to calcualte, $E_{var}$ really reduces to finding the sample variance of the log-transformed species abundances and then standardizing it to a value between 0 and 1.

```{r}
X <- var(log(RAC))
Evar <- 1 - (2/pi)*atan(X) # these operations make the value of Evar range between 0 and 1
Evar # print Evar for Site1
```

We can confirm this with a more explicit R chunk:

```{r}
P <- log(RAC) # log-tranform the abundances of the RAC and assign them to a vector P
AvgAb <- mean(P) # find the average of the log abundances
X <- 0 # assign zero to variable X
Evar <- 0 # declare a scalar varible Evar

for (x in P) {
  X = X + (x - AvgAb)^2 / (S - 1)
}

Evar = 1 - (2/pi)*atan(X) # these operations make the value of Evar range between 0 and 1
Evar # print Evar for Site1
```

** I need to go through this more carefully, but I think I'm missing the "takehome" point of the R chunk above. {JTL}**
As you can see, the value of $E_{var}$ suggests an intermediate value of species evenness. 
Likewise, you can see that different evenness metrics, even when bound between 0 and 1 still return very different values. 
This is not because each is inaccurate but because each metric emphasizes different aspects of the SAD (or RAC). 
Where Simpson's evenness focuses more greatly on dominant species, $E_{var}$ places 

is nearly as influenced by the abundance of dominant species as that of rare ones. 
In effect, the metric of evenness one used, even if only using NHC, depends on the aspects of similarity in species abundances one prefers to emphasize. ** I don't understand wording here...be clearer{JTL}**

## Species Diversity
You saw in the derivation of Shannon's evenness and Simpson's evenness that we called the R diversity function, and then specified either "shannon" or "simp". 
In effect, in estimating Shannon's or Simpson's evenness we first estimated Shannon's diversity (*H'*) and Simpson's diversity (*D*)! Species diversity is commonly defined as the relationship between species richness and species evenness, often of a form similar to ${Diversity = } \frac{Evenness} {Richness}$. 
Can you see this in looking back at how we calculated Shannon's and Simpson's evenness? **How will we know whether or not students "see" this?**
Here, we will estimate popular indices of diversity (Shannon's, Simpson's, Fisher's alpha) using our own derivation and then check this against vegan's estimates. **"For the purpose of..." is the learning objective to just check vegan to make sure it's right?**

### Shannon's diversity (or entropy)
Shannon's diversity metric is really just Shannon's information entropy, a measure of uncertainty. 
This metric is calculated as ${H' = -}  \sum p_i ln(p_i)$. 
Let's calculate Shannon's diversity for Site1 and then compare it to vegan's estimate:

```{r}
H <- 0
for (sp in RAC){
  p = sp / sum(RAC)
  H = H - p*log(p) 
}
H

diversity(RAC, index="shannon")
```

**Are they the same? {MEM}**

### Simpson's diversity (or dominance)
Simpsons diversity is a straightforward metric and is calculated as ${D = } \sum p_i^2$ where $p_i$ is the proportion of individuals found in the *i*th species. 
Simpson's index is often expressed as 1/*D* or 1-*D*, so that diversity naturally increases with 1/*D*. 
Let's calculate Simpson's diversity for Site1 and then compare it to vegan's estimate:

```{r}
D <- 0.0
N <- sum(RAC)
for (ni in RAC){
  D = D + (ni*ni)/(N*N)
}

invD <- 1/D # using the 1/D variation
invD

invD <- diversity(RAC, "inv") # using vegan
invD

D <- 1 - D # using the 1 - D variation
D

D <- diversity(RAC, "simp") # using vegan
D
```

### Fisher's $\boldsymbol\alpha$
R.A. Fisher (1943) derived one of the first and most successful models for how abundance varies among species, i.e., the log-series distribution. 
This model has only a single fitted parameter, i.e., $\alpha$, which is implicitly defined by the log-series model: ${S =} \alpha * ln(1 + n/\alpha)$. 
However, this does not tell us exactly how to estimate $\alpha$, and this is because $\alpha$ is a fitted parameter. 
Later, we will explore the log-series. 
For now, let's concentrate on the Fisher's $\alpha$, which has often been used as a diversity metric and is the root of 'alpha' diversity. 
Fisher's $\alpha$, according to the authors of the vegan package is asymptotically similar to inverse Simpson's. 
Let's compare using Site1.

```{r}
invD <- diversity(RAC, "inv")
invD
Fisher <- fisher.alpha(RAC)
Fisher
```

As we can see, the two measurements are somewhat similar. 
They would converge if our community was much greater in total abundance and richness. 
However, discussion of Fisher's $\alpha$ introduces a new concept, that is, of estimating diversity instead of just calculating a diversity metric. 
The difference being that an estimate of diversity implicitly or explicitly accounts for samplign error, that is, the fact that when samplign most ecological communities that we are not observing every single individual.

## Estimating Richness and Diversity



### Other Diversity Estimators
**We need to think about how this will fit in with the flow of the rest of the exercise {JTL}**

___

## Homework

1) As stated by Magurran (2004) the  ${D = } \sum p_i^2$ derivation of Simpson's D really only applies to communities of infinite size. 
For anything but an infinitely large community, Simpson's index is calculated as ${D = } \sum \frac{n_i(n_i -1)} {N(N-1)}$. 
Calculate Simpson's D, 1 - D, and Simpson's inverse (i.e. 1/D) for Site1.

2) ...

3) use knitr to create a pdf, push it to GitHub, and create a pull request

**We have talked about using the Shaw et al. 2008 paper as a model for a homework assignment.  See [here](https://quantitativebiodiversity.etherpad.mozilla.org/4) {JTL}**

